---
title: ""
output:
  html_document:
    theme: united
    toc: yes
    toc_depth: 4
    toc_float: true
    code_folding: hide
  pdf_document:
    toc: yes
bibliography: references.bib

---

```{r setup, echo=FALSE, warning=FALSE, message=FALSE}
clientname="Margaret Stefater"
clientemail="margaret.stefater@childrens.harvard.edu"
labPI="Nicolas Stylopolous"
lablocation="Boston Children's Hospital"
analystname="John Hutchinson"
analystemail="jhutchin@hsph.harvard.edu"

library(knitr)

opts_chunk$set(warning=FALSE, error=FALSE, message=FALSE, echo=TRUE,cache=TRUE, tidy.opts=list(keep.blank.line=FALSE, width.cutoff=120), dev="svg")
options(width=200)
```

---

Array analysis for `r clientname` (`r clientemail`), `r labPI` group at `r lablocation`.  

Contact `r analystname` (`r analystemail`) for additional details.

The most recent update of this html document occurred: `r date()`

The sections below provide code to reproduce the included results and plots. 

---

# Methods Summary  

Initially, all Affymetrix HTA 2.0 arrays were processed using the 'oligo' BioConductor package [@Carvalho2010] and quality-controlled with arrayQualityMetrics [@Kauffmann2009]. After exclusion of clear outliers, the remaining Affymetrix HTA 2.0 arrays were processed using the TAC4.0 software from Affymetrix/Thermo Fisher using their RMA-SST algorithm and imported back into this workflow. Differentially expressed genes were identified using limma [@Smyth2005].

---

# Setup

## Variables
Working directories, files and other variables necessary to the analysis.

```{r variables}
## Setup Data and Results directory variables
#baseDir <- file.path("/Volumes/orch/PIs/nicholas_stylopoulos/ns_microarray_covariate/")
baseDir <- file.path("~/Orchestra/PIs/nicholas_stylopoulos/ns_microarray_covariate/")

dataDir <- "~/Work/HBC/projects/ns_microarray_covariate/data/"
metaDir <- file.path(baseDir, "metadata")
resultsDir <- file.path(baseDir, "results")
firstroundresultsDir <- "~/Dropbox (HBC)/HBC Team Folder (1)/Consults/nicholas_stylopolous/ns_microarray_covariate/20171115 Archive/results/"
setwd(baseDir)

cbPalette <- c("#999999", "#E69F00", "#56B4E9", "#009E73", "#F0E442", "#0072B2", "#D55E00", "#CC79A7") # colorblind friendly palette
covarsfilename="covars.desc" # tab delimited file describing samples
lowintensity.percentile=0.1
mad.quantile.cutoff=0.1
pvalue.cutoff=0.05
highlight.color="green"
lfc.cutoff=1
```

## Libraries

[Bioconductor](http://www.bioconductor.org) and [R](http://cran.r-project.org/) libraries used to process and visualize the data.

```{r libraries_variables, results='hide'}
library(knitr) # for simple tables
library(oligo) # array utilities
library(limma) # array statistical analyses
library(CHBUtils) # some homegrown functions
library(reshape2) # data format utility
library(ggplot2) # pretty graphs
library(ggdendro) # for pretty dendrograms
library(pheatmap) # pretty heatmaps
library(dplyr) # data format utility 
library(hta20transcriptcluster.db) #new package for annotation
library(affxparser)

library(rio)
library(tidyr)
library(janitor)
library(plyr)
library(DT)
library(DEGreport)
library(treemap)
library(gProfileR)
library(WGCNA)
```


---

# Import Data and Metadata

- load in phenotypes and array names from metadata file (covars.desc) in "metadata" directory
  - this file contains the names and descriptions of CEL files contained in the data directory 

### Create new expression set objects to normalize

```{r dataload, results='hide'}
chpfiles <- list.files(dataDir, pattern="*.chp")

# Load CHP files in to dataframe with named columns
setwd(dataDir)
chpdata <- laply(chpfiles, function(chpfilename) {
  chpdata <- readChp(chpfilename)
  probeids <- chpdata$QuantificationEntries$ProbeSetName
  probequants <- as.numeric(chpdata$QuantificationEntries$QuantificationValue)
  quantdata <- cbind(probeids, probequants) %>% as.data.frame()
  filestub <- strsplit(chpfilename, "_")[[1]][1]
  names(quantdata) <- c(paste(filestub, "probeid", sep="-"), filestub)
  return(list("quantdata"=quantdata))
})
chpdata <- do.call(cbind,chpdata) %>% as.data.frame()
probeids <- dplyr::select(chpdata, contains("probeid"))[,1]
chpdata <- dplyr::select(chpdata, -contains("probeid"))

chpdata <- chpdata %>% as.data.frame()

chpdata <- apply(chpdata, 2, as.numeric) %>% as.data.frame()
row.names(chpdata) <- probeids

affyNorm <- ExpressionSet(assayData=as.matrix(chpdata))

covars <- names(chpdata)
names(covars) <- names(chpdata)
covars <- as.data.frame(covars)
names(covars) <- "sampleId"
pData(affyNorm) <- covars
sampleNames(affyNorm) <- names(chpdata)
annotation(affyNorm) <- "pd.hta.2.0"
```

## Sample metadata

```{r covarsdisplay, results='asis'}
# Sample information table
kable(pData(affyNorm))
```

---

## Annotate

So far we have only been working with the probesets, without reference to the genes they assay. Here we load in metadata about the probesets on the array (feature data), the gene symbols in particular. WE get probeids and transcript information from NetAffx using the NetAffx BioConductor package and the gene symbols and entrezIDs from the BioConductor annotation packages.

```{r features, results='hide'}
featureData(affyNorm) <- getNetAffx(affyNorm, "transcript") # this will load the Affymetrix annotation, including the probeID, into the fData
# get gene symbols and entrezIDs for all probesets
featureData(affyNorm)$symbol <- as.character(unlist(mget(featureNames(affyNorm), hta20transcriptclusterSYMBOL, ifnotfound=NA))) # curated annotations from Bioconductor 
featureData(affyNorm)$entrezID <- as.character(unlist(mget(featureNames(affyNorm), hta20transcriptclusterENTREZID, ifnotfound=NA))) # curated annotations from Bioconductor 
```

## Filter Probesets
Reducing the number of genes assayed reduces the multiple test correction and may allow us to identify more differentially expressed genes.

Here we just filter to the same probes as in the first analysis. (using results from 20171204 file on Dropbox)

```{r filtertomatch, echo=TRUE}
library(rio)
dat1 <- import(file.path(firstroundresultsDir, "stats_exprs_baseline_vs_one_month.xlsx"))
dat2 <- import(file.path(firstroundresultsDir, "stats_exprs_baseline_vs_six_month.xlsx"))
dat3 <- import(file.path(firstroundresultsDir, "stats_exprs_one_month_vs_six_month.xlsx"))

identical(sort(dat1$probesetid),sort(dat2$probesetid))
identical(sort(dat1$probesetid),sort(dat3$probesetid))
identical(sort(dat2$probesetid),sort(dat3$probesetid))

probesets <- dat1$probesetid
affyNorm_sub <- affyNorm[which(featureNames(affyNorm) %in% probesets),]

intensities <- exprs(affyNorm_sub)
annots <- fData(affyNorm_sub)

output <- cbind(intensities, annots)

export(output, file.path(resultsDir,"secondset_annotated_rmasst_normalized_intensities.xlsx"))
```

[Annotated RMASST normalized intensities from second set of samples](./results/secondset_annotated_rmasst_normalized_intensities.xlsx)

---

# R Session Info

(useful if replicating these results)

```{r sessioninfo}
sessionInfo()
```

---

# References



