---
title: "Differential Expression - OCP pairwise comparisons with RUVseq batch correction"
author: "`r getOption('author')`"
date: "`r Sys.Date()`"
bibliography: bibliography.bib
params:
    bcbFile: "data/2018-04-12/bcb_sub.rda"
    design: !r formula(~W1+W2+W3+ treatment)
    contrast: !r c("treatment", "CTL", "BOTH")
    alpha: 0.05
    lfc: 0
    outputDir: "."
    ruv: "data/2018-04-12/ruv.rda"
---

```{r setup, message=FALSE, warning=FALSE}
library(DESeq2)
library(bcbioRNASeq)
library(tidyverse)
library(knitr)
library(rio)

# Shared RMarkdown settings
prepareRNASeqTemplate()
if (file.exists("setup.R")) {
    source("setup.R")
}

# Load bcbioRNASeq object
bcbName <- load(params$bcbFile)
bcb <- get(bcbName, inherits = FALSE)

ruvName <- load(params$ruv)
ruv <- get(ruvName, inherits = FALSE)
ruv_counts <- ruv$normalizedCounts
ruv_counts <- ruv_counts[,row.names(colData(bcb))]
# add ruvseq based adjustment factors to colData
colData(bcb)$W1 <- ruv$W[,"W_1"]
colData(bcb)$W2 <- ruv$W[,"W_2"]
colData(bcb)$W3 <- ruv$W[,"W_3"]

# subset everything to only OCP samples
ruv_counts <- ruv_counts[,which(colData(bcb)$tissue=="OCP")]
bcb <- bcb[,which(colData(bcb)$tissue=="OCP")]

#drop unnecessary replicates - default to keeping replicate with most mapped reads
mycoldata <- as.data.frame(colData(bcb))
metrics <- metadata(bcb)$metrics %>% tibble::rownames_to_column("sampleName")
mycoldata <- inner_join(mycoldata,metrics)
singletids <- mycoldata %>% group_by(sampleid) %>% filter(mappedReads == max(mappedReads)) %>% ungroup() %>% dplyr::select(sampleName) %>% unlist() %>% as.vector()

ruv_counts <- ruv_counts[,which(mycoldata$sampleName %in% singletids)]
bcb <- bcb[,which(mycoldata$sampleName %in% singletids)]

# reorder by CTL, ZA, GCSF and BOTH
bcb <- bcb[,c(6,7,13,14,15,16,17,8,9,10,11,12,1,2,3,4,5)]
ruv_counts <- ruv_counts[,c(6,7,13,14,15,16,17,8,9,10,11,12,1,2,3,4,5)]

# add in columns for regression
temp <- colData(bcb)
temp$gcsf <- ifelse(temp$treatment=="CTL", "NO", ifelse(temp$treatment=="ZA","NO", "YES"))
temp$za <- ifelse(temp$treatment=="CTL", "NO", ifelse(temp$treatment=="GCSF","NO" , "YES"))
colData(bcb) <- temp

# Design formula
# help("design", "DESeq2")
design <- params$design

# Contrast
# help("results", "DESeq2")
#   1. Design matrix parameter.
#   2. Numerator for LFC (expt).
#   3. Denominator for LFC (control).
contrast <- params$contrast

# Significance cutoffs
alpha <- params$alpha
lfc <- params$lfc

# Directory paths
outputDir <- params$outputDir
countsDir <- file.path(outputDir, "results", "counts")
deDir <- file.path(outputDir, "results", "differential_expression")

source("functions.R")
```

```{r header, child="_header.Rmd", eval=file.exists("_header.Rmd")}
```

```{r dds, results="hide", warning=FALSE, message=FALSE}
# have to drop the coldata out to a dataframe as colData(bcb) will only return factors, and deseq needs numeric values
mycoldata <- colData(bcb)
mycoldata$W1 <- mycoldata$W1 %>% as.character() %>% as.numeric()
mycoldata$W2 <- mycoldata$W2 %>% as.character() %>% as.numeric()
mycoldata$W3 <- mycoldata$W3 %>% as.character() %>% as.numeric()
dds <- DESeqDataSetFromTximport(
    txi = bcbio(bcb, "tximport"),
    colData = mycoldata,
    design = design) 
dds$treatment <- relevel(dds$treatment, "CTL")

dds <- DESeq(dds)
rld <- rlog(dds)
```

# Alpha level (FDR) cutoffs {.tabset}

Let's take a look at the number of genes we get with different false discovery rate (FDR) cutoffs. These tests subset *P* values that have been multiple test corrected using the Benjamini Hochberg (BH) method [@Benjamini:1995ws].

## GCSF versus CTL

```{r alpha_summary_gcsf, results="asis"}
alphaSummary(dds, contrast=c("treatment", "GCSF", "CTL"))
```

## ZA versus CTL

```{r alpha_summary_za, results="asis"}
alphaSummary(dds, contrast=c("treatment", "ZA", "CTL"))
```

## BOTH versus CTL

```{r alpha_summary_both, results="asis"}
alphaSummary(dds, contrast=c("treatment", "BOTH", "CTL"))
```


# Results

```{r res_gcsf}
contrast=c("treatment", "GCSF", "CTL")
resUnshrunken <- results(
    dds,
    contrast = contrast,
    alpha = alpha)

# DESeqResults with shrunken log2 fold changes (LFC)
# help("lfcShrink", "DESeq2")
# Only `coef` or `contrast` can be specified, not both
# Use the correct `coef` number to modify from `resultsNames(dds)`
resShrunken <- lfcShrink(
    dds = dds,
    # coef = 2,
    contrast = contrast,
    res = resUnshrunken)

# Use shrunken LFC values by default
res_gcsf <- resShrunken
```

```{r res_za}
contrast=c("treatment", "ZA", "CTL")
resUnshrunken <- results(
    dds,
    contrast = contrast,
    alpha = alpha)

# DESeqResults with shrunken log2 fold changes (LFC)
# help("lfcShrink", "DESeq2")
# Only `coef` or `contrast` can be specified, not both
# Use the correct `coef` number to modify from `resultsNames(dds)`
resShrunken <- lfcShrink(
    dds = dds,
    # coef = 2,
    contrast = contrast,
    res = resUnshrunken)

# Use shrunken LFC values by default
res_za <- resShrunken
```

```{r res_both}
contrast=c("treatment", "BOTH", "CTL")
resUnshrunken <- results(
    dds,
    contrast = contrast,
    alpha = alpha)

# DESeqResults with shrunken log2 fold changes (LFC)
# help("lfcShrink", "DESeq2")
# Only `coef` or `contrast` can be specified, not both
# Use the correct `coef` number to modify from `resultsNames(dds)`
resShrunken <- lfcShrink(
    dds = dds,
    # coef = 2,
    contrast = contrast,
    res = resUnshrunken)

# Use shrunken LFC values by default
res_both <- resShrunken
```

We performed the analysis using a BH adjusted *P* value cutoff of `r alpha` and a log fold-change (LFC) ratio cutoff of `r lfc`.

## Plots 

### Mean average (MA) {.tabset}

An MA plot compares transformed counts on `M` (log ratio) and `A` (mean average) scales [@Yang:2002ty].

#### GCSF versus CTL

```{r plot_ma_gcsf}
plotMA(res_gcsf)
```

#### ZA versus CTL

```{r plot_ma_za}
plotMA(res_za)
```

#### BOTH versus CTL

```{r plot_ma_both}
plotMA(res_both)
```

### Volcano {.tabset}

`r lfc=1`

A volcano plot compares significance (BH-adjusted *P* value) against fold change (log2) [@Cui:2003kh; @Li:2014fv]. Genes hightlighted in purpble have an adjusted *P* value less than `r alpha` and an absolute log2FoldChange of more than `r lfc` and are likely to be the top candidate genes of interest. 

#### GCSF versus CTL

```{r plot_volcano_gcsf}
plotVolcanoJH(res_gcsf, lfc=lfc, alpha = alpha, color="purple", padj=TRUE, alphashade = 0.5, pointsize=3)
```

#### ZA versus CTL

```{r plot_volcano_za}
plotVolcanoJH(res_za, lfc = lfc,alpha = alpha, color="purple", padj=TRUE, alphashade = 0.5, pointsize=3)
```

#### BOTH versus CTL

```{r plot_volcano_both}
plotVolcanoJH(res_both, lfc = lfc,alpha = alpha, color="purple", padj=TRUE, alphashade = 0.5, pointsize=3)
```

`r lfc=0`


## Top tables of differentially expressed genes {.tabset}

```{r results_tables, results="hide"}
resTbl_gcsf <- resultsTables(res_gcsf, lfc = lfc, dir = deDir, annotable=metadata(bcb)$annotable, write=TRUE)
resTbl_za <- resultsTables(res_za, lfc = lfc, dir = deDir, annotable=metadata(bcb)$annotable, write=TRUE)
resTbl_both <- resultsTables(res_both, lfc = lfc, dir = deDir, annotable=metadata(bcb)$annotable, write=TRUE)
```

DEG tables are sorted by BH-adjusted P value, and contain the following columns:

- `ensgene`: [Ensembl][] gene identifier.
- `baseMean`: Mean of the normalized counts per gene for all samples.
- `log2FoldChange`: log2 fold change.
- `lfcSE`: log2 standard error.
- `stat`: Wald statistic.
- `pvalue`: Walt test *P* value.
- `padj`: BH adjusted Wald test *P* value (corrected for multiple comparisons; aka FDR).
- `externalGeneName`: [Ensembl][] name (a.k.a. symbol).
- `description`: [Ensembl][] description.
- `geneBiotype`: [Ensembl][] biotype (e.g. `protein_coding`).

Only the top 50 up- and down-regulated genes (arranged by log2 fold change for both up and down regulated genes) are shown.

### GCSF versus CTL

```{r top_tables_gcsf, results="asis"}
topTables(resTbl_gcsf)
```

### ZA versus CTL

```{r top_tables_za, results="asis"}
topTables(resTbl_za)
```

### BOTH versus CTL 

```{r top_tables_both, results="asis"}
topTables(resTbl_both)
```

## Overlaps of differentially expressed genes in pairwise comparisons

I show the overlaps in differentially expressed genes using two different methods. 

a) classic Venn/Euler diagram - possibly easier to understand but less accurate and less information
b) using an UpSet plot to show all set sizes and sizes for all unique intersections. Read more about them here: https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4720993/

```{r venns, fig.height=5}
sig_both <-   res_both %>% as.data.frame() %>% rownames_to_column(var="ensgene") %>% filter(padj<alpha) %>% dplyr::select(., ensgene) %>% unlist() %>% as.vector()
sig_gcsf <-   res_gcsf %>% as.data.frame() %>% rownames_to_column(var="ensgene") %>% filter(padj<alpha) %>% dplyr::select(., ensgene) %>% unlist() %>% as.vector()
sig_za <-   res_za %>% as.data.frame() %>% rownames_to_column(var="ensgene") %>% filter(padj<alpha) %>% dplyr::select(., ensgene) %>% unlist() %>% as.vector()
sig_both_gcsf <- intersect(sig_both, sig_gcsf)
sig_both_za <- intersect(sig_both, sig_za)
sig_gcsf_za <- intersect(sig_gcsf, sig_za)
sig_both_gcsf_za <- intersect(intersect(sig_gcsf, sig_za), sig_both)

 num_both <- length(sig_both)
 num_gcsf <- length(sig_gcsf)
 num_za <- length(sig_za)
num_both_gcsf <- length(sig_both_gcsf)
num_both_za <- length(sig_both_za)
num_gcsf_za <- length(sig_gcsf_za)
num_both_gcsf_za <- length(sig_both_gcsf_za)


print(num_gcsf)

library(eulerr)
fit <- euler(c("GCSF" = num_gcsf, "ZA" = num_za, "BOTH"= num_both, "GCSF&ZA" = num_gcsf_za, "GCSF&BOTH"=num_both_gcsf, "BOTH&ZA"=num_both_za, "GCSF&ZA&BOTH"=num_both_gcsf_za  ), input="union")

plot(fit, alpha=0.6,
     fill=c("#FF0000","#0000FF", "#FFFF00" ),
     border="grey", quantities=TRUE)
```





# File downloads

The results are saved as gzip-compressed comma separated values (CSV). Gzip compression is natively supported on [macOS][] and Linux-based operating systems. If you're running Windows, we recommend installing [7-Zip][]. CSV files can be opened in [Excel][] or [RStudio][].


## Count matrices

- [`normalizedCounts.csv.gz`](`r file.path(countsDir, "normalizedCounts.csv.gz")`): Use to evaluate individual genes and/or generate plots. These counts are normalized for the variation in sequencing depth across samples.
- [`tpm.csv.gz`](`r file.path(countsDir, "tpm.csv.gz")`): Transcripts per million, scaled by length and also suitable for plotting.
- [`rawCounts.csv.gz`](`r file.path(countsDir, "rawCounts.csv.gz")`): Only use to perform a new differential expression analysis. These counts will vary across samples due to differences in sequencing depth, and have not been normalized. Do not use this file for plotting genes.

## Differentially expressed genes (DEG)

All files can be found in this [folder](./results/differential_expression/). 

\*all.csv.gz - statistics and annotations for all genes       
\*deg.csv.gz - statistics and annotations for differentially expressed genes
\*deg_lfc_down.csv.gz - statistics and annotations for differentially expressed genes that are up in first listed sample class (i.e. if filename is 1_vs_2, genes are more highly expressed in 1)
\*deg_lfc_up.csv.gz - statistics and annotations for differentially expressed genes that are down in first listed sample class (i.e. if filename is 1_vs_2, genes are more highly expressed in 1)

DEG tables are sorted by BH-adjusted P value, and contain the following columns:

- `ensgene`: [Ensembl][] gene identifier.
- `baseMean`: Mean of the normalized counts per gene for all samples.
- `log2FoldChange`: log2 fold change.
- `lfcSE`: log2 standard error.
- `stat`: Wald statistic.
- `pvalue`: Walt test *P* value.
- `padj`: BH adjusted Wald test *P* value (corrected for multiple comparisons; aka FDR).
- `externalGeneName`: [Ensembl][] name (a.k.a. symbol).
- `description`: [Ensembl][] description.
- `geneBiotype`: [Ensembl][] biotype (e.g. `protein_coding`).



```{r footer, child="_footer.Rmd", eval=file.exists("_footer.Rmd")}
```
