---
title: "Quality control report"
date: "`r BiocStyle::doc_date()`"
author: "Michael J. Steinbaugh"
bibliography: bibliography.bib
---

```{r setup, child="setup.Rmd"}
```

```{r header, child="header.Rmd"}
```

```{r bcbioRnaseqData, include=FALSE}
rnaseq <- bcbioRnaseqData(project)
save(rnaseq, file = "data/rnaseq.rda")
```

```{r deseq, results="hide"}
# Import counts from `sailfish`
txi <- bcbioSailfish(project, summary)

# DESeqDataSet
design <- stats::formula(~ group)
dds <- txi %>%
    DESeq2::DESeqDataSetFromTximport(colData = metadata,
                                     design = design) %>%
    # Can optionally evaluate `betaPrior = TRUE/FALSE`
    DESeq2::DESeq(.)
save(dds, file = "data/dds.rda")

# DESeqTransform
rld <- DESeq2::rlog(dds)
vsd <- DESeq2::varianceStabilizingTransformation(dds)
save(rld, vsd, file = "data/dt.rda")
```

```{r deseq_qc, eval=FALSE, include=FALSE}
DESeq2::resultsNames(dds) %>% print
DESeq2::sizeFactors(dds) %>% print

# The following plot shows the dispersion by mean of normalized counts. We
# expect the dispersion to decrease as the mean of normalized counts increases.
DESeq2::plotDispEsts(dds)

# Several quality metrics are first assessed to explore the fit of the model,
# before differential expression analysis will be performed. The plots below
# show the standard deviation of normalized counts (`normalized_counts`) using
# `log2()`, `rlog()`, and variance stabilizing (`vst()`) transformations by
# `rank(mean)`. The transformations greatly reduce the standard deviation, with
# `rlog` generally stabilizing the variance best across the mean.
vsn::meanSdPlot(log2(normalized_counts + 1))
vsn::meanSdPlot(assay(vsd))
vsn::meanSdPlot(assay(rld))
```

```{r tpm}
# Transcripts per million
tpm <- txi$abundance
save(tpm, file = "data/tpm.rda")
write.csv(tpm, file = "results/tpm.csv")
```

```{r counts}
raw_counts <- DESeq2::counts(dds)
raw_counts_nonzero <- raw_counts %>%
    .[rowSums(.) > 0, ]
normalized_counts <- DESeq2::counts(dds, normalized = TRUE)
normalized_counts_nonzero <- normalized_counts %>%
    .[rowSums(.) > 0, ]
tmm_normalized_counts <- tmmNormalize(raw_counts)
save(raw_counts,
     raw_counts_nonzero,
     normalized_counts,
     normalized_counts_nonzero,
     tmm_normalized_counts,
     file = "data/counts.rda")
write.csv(raw_counts,
          file = "results/raw_counts.csv")
write.csv(normalized_counts,
          file = "results/normalized_counts.csv")
```



# Read metrics

## Total reads

We advise that samples should preferably have above 20 million reads and at least 10 million reads. All of the samples look good.

```{r totalReads, fig.width=6, fig.height=9}
totalReads(summary)
```


## Mapped reads

The number of mapped reads should correspond fairly closely to the number of total reads.

```{r mappedReads, fig.width=6, fig.height=9}
mappedReads(summary)
```


## Mapping rate

The mapping rate represents the percentage of reads that successfully mapped to the reference genome. Low mapping rates are indicative of sample contamination, poor sequencing quality, or other artifacts. All of the samples have an acceptable mapping rate.

```{r mappingRate, fig.width=6, fig.height=9}
mappingRate(summary)
```


## Number of genes detected

We're seeing detection above 20k genes for all replicates, which is normal for mouse samples.

```{r genesDetected, fig.width=6, fig.height=9}
genesDetected(raw_counts, summary)
```



## Gene detection saturation

Gene detection saturation was nearly reached for the sample input, which means we're able to detect the majority of expressed genes.

```{r saturation}
geneDetectionSaturation(raw_counts, summary)
```


## Exonic mapping rate

Good quality samples should have an exonic mapping rate of at least 60%. We're seeing a relatively low mapping rate here, which isn't desirable.

```{r exonicMappingRate, fig.width=6, fig.height=9}
exonicMappingRate(summary)
```


## Intronic mapping rate

The low RNA input (< 20ng) could be responsible for us seeing a relatively high intronic mapping rate.

```{r intronicMappingRate, fig.width=6, fig.height=9}
intronicMappingRate(summary)
```


## rRNA mapping rate

Without poly(A) enrichment (preferred for mRNA analysis) or ribodepletion (preferred for ncRNA analysis), ~80% of sample reads will map to rRNA. Good quailty samples should have rRNA mapping rate below 10%, otherwise the enrichment step may not have worked very well, and the depth of informative reads will be lower. All the samples look clean.

```{r rrnaMappingRate, fig.width=6, fig.height=9}
rrnaMappingRate(summary)
```


## Counts per gene

Generally, we expect similar count spreads for all genes between samples unless the library sizes or total RNA expression are different.

The log10 TMM-normalized counts per gene normalization method [@robinson2010tmm] equates the overall expression levels of genes between samples under the assumption that the majority of them are not differentially expressed. Therefore, by normalizing for total RNA expression by sample, we expect the spread of the log10 TMM-normalized counts per gene to be similar for every sample.

All of the samples have similar distributions, which is good.

```{r countsPerGene, fig.width=6, fig.height=9}
countsPerGene(raw_counts, metadata)
countsPerGene(tmm_normalized_counts, metadata)
```


## Count density

Generally, we expect similar count spreads for all genes between samples unless the total expressed RNA per sample is different.

```{r countDensity}
countDensity(tmm_normalized_counts, metadata)
```



# Sample similarity analysis

Before performing similarity analysis, we transform counts to log2, which acts to minimize large differences in sequencing depth and helps normalize all samples to a similar dynamic range. For RNA-Seq count data, variance increases with the mean. Logarithmic transformation of normalized count values with a small pseudocount will account for large variations seen between the highest expressing genes so that these genes won't dominate the PCA plots. However, due to the strong noise among low count values due to Poisson, the general log2 transformation will amplify this noise, and instead, low count genes will now dominate the PCA plots. So instead, we use a transformation (`rlog`) that gives similar results for high counts as a log2 transformation but also shrinks the values of low counts towards the genesâ€™ average across samples. We do this with the `rlog` function in the [`DESeq2`][] package [@love2014deseq2], which we will later use for differential gene expression analysis.


## Principal component analysis (PCA)

PCA [@jolliffe2002pca] is a multivariate technique that allows us to summarize the systematic patterns of variations in the data. PCA takes the expression levels for genes and transforms it in principal component space, reducing each sample into one point. Thereby, we can separate samples by expression variation, and identify potential sample outliers. The PCA plot is a way to look at how samples are clustering.

Here we see clear clustering of the lanesplit technical replicates, which is good. However, the samples are not clustering well by treatment group (`shapes`), which likely indicates that there are not large changes in gene expression due to the cage enrichment.

```{r deseqPca}
deseqPca(rld, factor = factor[1:3])
```



## Hierarchical clustering

Inter-correlation analysis (ICA) is another way to look at how well samples cluster by plotting the correlation between the expression profiles of the samples.

Again, we see good clustering of the technical replicates, but the grouping by treatment isn't that strong. Ideally, you'd like to see all the `EE` and `SH` samples groups grouped together.

```{r corHeatmap, fig.width=11, fig.height=8.5}
corHeatmap(rld, metadata, factor = factor)
```


# Gender markers

Finally, we can evaluate the expression counts of gender-specific markers [@karyagyna2010gender; @staedtler2013gender] to blindly determine gender. All of the mice appear to be male, as expected.

```{r genderMarkers}
genderMarkers(tpm, "mmusculus")
```



# Summary

Overall, the samples have an acceptable sequencing depth and mapping efficiency. We're not seeing clear clustering of the samples based on the treatment, but this may not be surprising if big changes in gene expression are not expected. I will proceed to perform differential expression analysis and functional enrichment on these samples.

```{r footer, child="footer.Rmd"}
```
