---
bibliography: "references.bib"
csl: "bioinformatics.csl"
title: "Anandasabpathy - Vaccina vaccination route effect on gene expression in immune derived cells"
output:
  html_document:
    code_folding: hide
    theme: yeti
    toc: yes
    toc_float: true
    toc_depth: 3
---

```{r setup, echo=FALSE, warning=FALSE, message=FALSE}
library("knitcitations")
cleanbib()
options("citation_format" = "pandoc")

clientname="Niroshana Anandasabapathy"
clientemail="niroananda@gmail.com"
labPI="Niroshana Anandasabapathy"
lablocation="BWH"
analystname="John Hutchinson"
analystemail="jhutchin@hsph.harvard.edu"

library(knitr)
opts_chunk$set(warning=FALSE, error=FALSE, message=FALSE, cache=TRUE, tidy.opts=list(keep.blank.line=FALSE, width.cutoff=120), dev="svg")
options(width=200)
```

---

Array analysis for `r clientname` (`r clientemail`), `r labPI` group at `r lablocation`.  

Contact `r analystname` (`r analystemail`) for additional details.

The most recent update of this html document occurred: `r date()`

The sections below provide code to reproduce the included results and plots. 

---

# Methods Summary  

All MoGene 2.0 ST arrays were processed using the 'oligo' BioConductor package `r citep("10.1093/bioinformatics/btq431")`, quality-controlled with arrayQualityMetrics `r citep("10.1093/bioinformatics/btn647")`, normalized with RMA `r citep("10.1093/biostatistics/4.2.249")`. Differentially expressed genes were identified using limma `r citep("http://link.springer.com/chapter/10.1007%2F0-387-29362-0_23")`.

---

# Setup

## Variables
Working directories, files and other variables necessary to the analysis.

```{r variables}
## Setup Data and Results directory variables
if(file.exists("/Users/johnhutchinson/Work/projects/na_vaccinia_scar")) {
  baseDir <- "/Users/johnhutchinson/Work/projects/na_vaccinia_scar"
} else {
  baseDir <- "~/projects/na_vaccinia_scar"
}
dataDir <- file.path(baseDir, "data")
metaDir <- file.path(baseDir, "meta")
resultsDir <- file.path(baseDir, "results")

cbPalette <- c("#999999", "#E69F00", "#56B4E9", "#009E73", "#F0E442", "#0072B2", "#D55E00", "#CC79A7") # colorblind friendly palette
covarsfilename="metadata.csv" # tab delimited file describing samples
lowintensity.percentile=0.1
mad.quantile.cutoff=0.1
pvalue.cutoff=0.05
highlight.color="green"
lfc.cutoff=0.58496251

pct_threshold <- 0.5
```

## Libraries

Bioconductor `r citep("10.1038/nmeth.3252")` and R `r citep("http://cran.r-project.org/")` libraries used to process the data.

```{r libraries}
library(arrayQualityMetrics) # array quality control reports
library(oligo) # array utilities

library(pd.mogene.2.0.st) # array layout annotation
library(mogene20sttranscriptcluster.db) # array probe to gene annotations

library(ggdendro)
library(CHBUtils)
library(limma)
library(ggplot2)
library(pheatmap) # pretty heatmaps

library(readr)
library(dplyr)
library(tidyr)
library(reshape2)

library(biomaRt)

library(pvca)


# use older version of biomart to allow homology changes
humanmart <- useDataset("hsapiens_gene_ensembl", useMart("ENSEMBL_MART_ENSEMBL", host="jul2016.archive.ensembl.org"))
mousemart <- useDataset("mmusculus_gene_ensembl", useMart("ENSEMBL_MART_ENSEMBL", host="jul2016.archive.ensembl.org"))

mouseattributes = listAttributes(mousemart)
mousefilters = listFilters(mousemart)

humanattributes = listAttributes(humanmart)
humanfilters = listFilters(humanmart)
```

## Functions
- some custom functions

```{r functions}
PCAplot.sd.eset <- function(eset=NULL,  title=NULL){
  eset.core <- exprs(eset)
  myPca.core <- prcomp(t(eset.core))
  # SD of components
  sdevdf <- data.frame(cbind(as.numeric(myPca.core$sdev),c(1:length(myPca.core$sdev))))
  sdevdf$prop <-  sdevdf$X1/sum(sdevdf$X1)
  sdevdf$cum <- cumsum(sdevdf$prop)
  ggplot(sdevdf, aes(x=X2, y=prop)) + 
    geom_point(size=4, color="red") + 
    scale_x_continuous('Component') + 
    scale_y_continuous('Standard Deviation') +
    ggtitle(title) +
    geom_line(data=sdevdf, aes(x=X2, y=cum))
}
heatmap_fn = function(data, metadata ,...) {
  pheatmap(data, annotation=metadata, clustering_method = "ward.D2",
           clustering_distance_cols = "correlation", ...)
}
```

---

# Import Data and Metadata

- load in phenotypes and array names from metadata file [metadata file](../meta/metadata.csv)
  - this file contains the names and descriptions of CEL files contained in the data directory 
- used these file names to load in the CEL files

```{r dataload, results='hide'}
covars <- read.csv(file.path(metaDir, covarsfilename)) ## simple comma separated  file with CEL file in second column  and sample metadata (i.e. sampleID, treatment group, batch etc.) in subsequent columns
covars <- covars %>% tbl_df() %>% 
  unite(., category, location, celltype, treatment, remove=FALSE) %>% 
  mutate(replicate = 1:n()) %>% 
  unite(., sampleid, category, replicate, remove=FALSE) %>% 
  unite(., celltype_location, celltype, location, sep="-", remove=FALSE) %>% 
  as.data.frame()

# this only works with the FULL path to the files
affyraw <- read.celfiles(file.path(dataDir, as.character(unlist(covars$celfile))))
pData(affyraw) <- covars 
sampleNames(affyraw) <- pData(affyraw)$array_code
validObject(affyraw)
```

## Sample metadata

```{r covars, results='asis'}
# Sample information table
output <- pData(affyraw)
row.names(output) <- as.character(seq(1:nrow(output)))
DT::datatable(output, rownames = FALSE)
```

---

# Pre-Processing and Qualilty Control

## Raw Data QC

- using arrayQualityMetrics library `r citep("10.1093/bioinformatics/btn647")`

```{r rawQC, eval=FALSE}
affyraw.eset <- ExpressionSet(assayData=exprs(affyraw))
pData(affyraw.eset) <- pData(affyraw)
data.prepped <- prepdata(affyraw.eset, do.logtransform = TRUE, intgroup=c("location","celltype", "treatment" ))

aqmheat <- aqm.heatmap(data.prepped)
aqmbox <- aqm.boxplot(data.prepped)
aqmdens <- aqm.density(data.prepped)
aqmma <- aqm.maplot(data.prepped)

aqm.writereport(list(aqmheat,  aqmbox, aqmdens, aqmma), outdir = file.path(resultsDir, "report_raw"), arrayTable = pData(affyraw.eset), reporttitle = "Raw QC Report")
```

**[Raw Data QC Report](../results/report_raw/index.html)**

Arrays #32, 33, 21 and 57 are outliers from the boxplots.
57, 62 and 5 also seem off by density plot.
Many of these also look bad by MA-plot.

Dropped `r kable(output[c(5,21,32,33,57,62),])`.

```{r dropoutliers}
outlierids <- pData(affyraw)$array_code[c(5,21,32,33,57,62)]
affyraw <- affyraw[,which(!pData(affyraw)$array_code %in% outlierids)]
```


## Background Correct and Normalize

- using RMA `r citep("10.1093/biostatistics/4.2.249")`

- summarize probesets on the gene ('core') level

```{r normalize, results='hide'}
affynorm.core <- oligo::rma(affyraw,  target="core", background=TRUE, normalize=TRUE)
```

## Normalized Data QC
- using arrayQualityMetrics library

```{r normQC, eval=FALSE}
data.prepped <- prepdata(affynorm.core, do.logtransform = FALSE, intgroup=c("location","celltype", "treatment" ))

aqmheat <- aqm.heatmap(data.prepped)
aqmbox <- aqm.boxplot(data.prepped)
aqmdens <- aqm.density(data.prepped)
aqmma <- aqm.maplot(data.prepped)

aqm.writereport(list(aqmheat,  aqmbox, aqmdens, aqmma), outdir = file.path(resultsDir, "report_rma_core"), arrayTable = pData(affynorm.core), reporttitle = "Normalized QC Report")
```

**[Normalized Data QC Report](../results/report_rma.core/index.html)**
 
Everything looks good now. You can start to see some separation between cDC and migDC derived cells.

---

# Unsupervised Clustering

## Correlation heatmaps of Normalized Intensities

- by cell type and location

The goal of these analyses are to naiively evaluate the variability within the raw data and determine whether this variability can predict the different sample groups

The first method produces a dendrogram by performing  
>  a hierarchical cluster analysis using a set of dissimilarities for the n objects being clustered

### CD8_plus-cDC {.tabset}

`r celltype="CD8_plus-cDC"`

```{r corheatmapprep.CD8_plus-c}
affynorm.sub <- affynorm.core[,which(pData(affynorm.core)$celltype_location==celltype)]
pd <- pData(affynorm.sub) %>% tbl_df %>% dplyr::select(treatment, batch, facsdate, RNA_prep_date) %>% as.data.frame()
```

#### Pearsons

```{r corheatmapprep.pearsonCD8_plus-c, fig.height=8}
heatmap_fn(data=cor(exprs(affynorm.sub), method="pearson"), metadata=pd)
```

#### Spearman

```{r corheatmapprep.spearman.CD8_plus-c, fig.height=8}
heatmap_fn(data=cor(exprs(affynorm.sub), method="spearman"), metadata=pd)
```

### CD103_plus-migDC  {.tabset}
 
```{r corheatmapprep.CD103_plus-migDC}
affynorm.sub <- affynorm.core[,which(pData(affynorm.core)$celltype_location==celltype)]
pd <- pData(affynorm.sub) %>% tbl_df %>% dplyr::select(treatment, batch, facsdate, RNA_prep_date) %>% as.data.frame()
```

#### Pearsons

```{r corheatmapprep.pearson.CD103_plus-migDC, fig.height=8}
heatmap_fn(data=cor(exprs(affynorm.sub), method="pearson"), metadata=pd)
```

#### Spearman

```{r corheatmapprep.spearman.CD103_plus-migDC, fig.height=8}
heatmap_fn(data=cor(exprs(affynorm.sub), method="spearman"), metadata=pd)
```

 
### CD11B_minus-migDC {.tabset}
 
 ```{r corheatmapprep.CD11B_minus-migDC}
affynorm.sub <- affynorm.core[,which(pData(affynorm.core)$celltype_location==celltype)]
pd <- pData(affynorm.sub) %>% tbl_df %>% dplyr::select(treatment, batch, facsdate, RNA_prep_date) %>% as.data.frame()
```

#### Pearsons

```{r corheatmapprep.pearson.CD11B_minus-migDC, fig.height=8}
heatmap_fn(data=cor(exprs(affynorm.sub), method="pearson"), metadata=pd)
```

#### Spearman

```{r corheatmapprep.spearman.CD11B_minus-migDC, fig.height=8}
heatmap_fn(data=cor(exprs(affynorm.sub), method="spearman"), metadata=pd)
```

### CD11B_plus-migDC {.tabset}
 
```{r corheatmapprep.CD11B_plus-migDC}
affynorm.sub <- affynorm.core[,which(pData(affynorm.core)$celltype_location==celltype)]
pd <- pData(affynorm.sub) %>% tbl_df %>% dplyr::select(treatment, batch, facsdate, RNA_prep_date) %>% as.data.frame()
```

#### Pearsons

```{r corheatmapprep.pearson.CD11B_plus-migDC, fig.height=8}
heatmap_fn(data=cor(exprs(affynorm.sub), method="pearson"), metadata=pd)
```

#### Spearman

```{r corheatmapprep.spearman.CD11B_plus-migDC, fig.height=8}
heatmap_fn(data=cor(exprs(affynorm.sub), method="spearman"), metadata=pd)
```

### CD11B_plus-cDC {.tabset}
 
```{r corheatmapprep.CD11B_plus-cDC}
affynorm.sub <- affynorm.core[,which(pData(affynorm.core)$celltype_location==celltype)]
pd <- pData(affynorm.sub) %>% tbl_df %>% dplyr::select(treatment, batch, facsdate, RNA_prep_date) %>% as.data.frame()
```

#### Pearsons

```{r corheatmapprep.pearson.CD11B_plus-cDC, fig.height=8}
heatmap_fn(data=cor(exprs(affynorm.sub), method="pearson"), metadata=pd)
```

#### Spearman

```{r corheatmapprep.spearman.CD11B_plus-cDC, fig.height=8}
heatmap_fn(data=cor(exprs(affynorm.sub), method="spearman"), metadata=pd)
```


# Exploring Potential Batch Effects
 - the data has a few potential sources of batch effects (RNA prep time, microarray batch, FACS date etc.) that could affect the outcome
 - we can explore whether any of these are major sources of variatoin using a technique called Principal Variance Component Analysis (PVCA)  as described in 
 `r citep("http://www.wiley.com/WileyCDA/WileyTitle/productCd-0470741384.html")`
 
 
### CD8_plus-cDC {.tabset}

`r celltype="CD8_plus-cDC"`

```{r, pvca_CD8_plus-cDC}
affynorm.sub <- affynorm.core[,which(pData(affynorm.core)$celltype_location==celltype)]
batch.factors <- c()
if (all(table(pData(affynorm.sub)$treatment)>1)){
  batch.factors <- c(batch.factors, "treatment")
}
if (all(table(pData(affynorm.sub)$batch)>1)){
  batch.factors <- c(batch.factors, "batch")
}
if (all(table(pData(affynorm.sub)$treatment)>1)){
  batch.factors <- c(batch.factors, "arraydate")
}
if (all(table(pData(affynorm.sub)$facsdate)>1)){
  batch.factors <- c(batch.factors, "facsdate")
}
if (all(table(pData(affynorm.sub)$RNA_prep_date)>1)){
  batch.factors <- c(batch.factors, "RNA_prep_date")
}
pvcaObj <- pvcaBatchAssess(affynorm.sub, batch.factors, pct_threshold)
plotme <- cbind(as.character(unlist(pvcaObj$label)), as.numeric(as.character(unlist(pvcaObj$dat)))) %>% tbl_df() %>% 
  set_colnames(c("label", "value")) %>% mutate(value=as.numeric(value))
ggplot(plotme, aes(x=label, y=value, fill=label))+geom_bar(stat="identity")
```


### CD103_plus-migDC {.tabset}

`r celltype="CD103_plus-migDC"`

```{r, pvca_CD103_plus-migDC, ref.label='pvca_CD8_plus-cDC'}
```

### CD11B_minus-migDC

`r celltype="CD11B_minus-migDC"`

```{r, pvca_CD11B_minus-migDC, ref.label='pvca_CD8_plus-cDC'}
```

### CD11B_plus-migDC

`r celltype="CD11B_plus-migDC"`

```{r, pvca_CD11B_plus-migDC, ref.label='pvca_CD8_plus-cDC'}
```

### CD11B_plus-cDC

`r celltype="CD11B_plus-cDC"`

```{r, pvca_CD11B_plus-cDC, ref.label='pvca_CD8_plus-cDC'}
```



---
# R Session Info

(useful if replicating these results)

```{r sessioninfo}
sessionInfo()
```


---

# References

```{r writebib, results='hide', echo=FALSE, message=FALSE, cache=FALSE}
write.bibtex(file="references.bib")
```