---
title: "Differential Expression - OCP run 1 pairwise comparisons, without pvalue cutoffs"
author: "`r getOption('author')`"
date: "`r Sys.Date()`"
bibliography: bibliography.bib
params:
    bcbFile: "data/2018-03-30/bcb_sub.rda"
    design: !r formula(~treatment)
    contrast: !r c("treatment", "CTL", "BOTH")
    alpha: 0.99999999999999
    lfc: 0.5849625
    outputDir: "."
---

```{r setup, message=FALSE}
library(DESeq2)
library(bcbioRNASeq)
library(CHBUtils)
library(tibble)

# Shared RMarkdown settings
prepareRNASeqTemplate()
if (file.exists("setup.R")) {
    source("setup.R")
}

# Load bcbioRNASeq object
bcbName <- load(params$bcbFile)
bcb <- get(bcbName, inherits = FALSE)

# Design formula
# help("design", "DESeq2")
design <- params$design

# Contrast
# help("results", "DESeq2")
#   1. Design matrix parameter.
#   2. Numerator for LFC (expt).
#   3. Denominator for LFC (control).
contrast <- params$contrast

# Significance cutoffs
alpha <- params$alpha
lfc <- params$lfc

# Directory paths
outputDir <- params$outputDir
countsDir <- file.path(outputDir, "results", "counts")
deDir <- file.path(outputDir, "results", "differential_expression")

source("functions.R")
```

```{r header, child="_header.Rmd", eval=file.exists("_header.Rmd")}
```

```{r dds, results="hide"}
treatment <- colData(bcb)$treatment

dds <- DESeqDataSetFromTximport(
    txi = txi(bcb),
    colData = colData(bcb),
    design = design) 
dds$treatment <- relevel(dds$treatment, "CTL")

dds <- DESeq(dds)
rld <- rlog(dds)
```


# Results

```{r res_gcsf}
contrast=c("treatment", "GCSF", "CTL")
resUnshrunken <- results(
    dds,
    contrast = contrast,
    alpha = alpha)

# DESeqResults with shrunken log2 fold changes (LFC)
# help("lfcShrink", "DESeq2")
# Only `coef` or `contrast` can be specified, not both
# Use the correct `coef` number to modify from `resultsNames(dds)`
resShrunken <- lfcShrink(
    dds = dds,
    # coef = 2,
    contrast = contrast,
    res = resUnshrunken)

# Use shrunken LFC values by default
res_gcsf <- resShrunken
```

```{r res_za}
contrast=c("treatment", "ZA", "CTL")
resUnshrunken <- results(
    dds,
    contrast = contrast,
    alpha = alpha)

# DESeqResults with shrunken log2 fold changes (LFC)
# help("lfcShrink", "DESeq2")
# Only `coef` or `contrast` can be specified, not both
# Use the correct `coef` number to modify from `resultsNames(dds)`
resShrunken <- lfcShrink(
    dds = dds,
    # coef = 2,
    contrast = contrast,
    res = resUnshrunken)

# Use shrunken LFC values by default
res_za <- resShrunken
```

```{r res_both}
contrast=c("treatment", "BOTH", "CTL")
resUnshrunken <- results(
    dds,
    contrast = contrast,
    alpha = alpha)

# DESeqResults with shrunken log2 fold changes (LFC)
# help("lfcShrink", "DESeq2")
# Only `coef` or `contrast` can be specified, not both
# Use the correct `coef` number to modify from `resultsNames(dds)`
resShrunken <- lfcShrink(
    dds = dds,
    # coef = 2,
    contrast = contrast,
    res = resUnshrunken)

# Use shrunken LFC values by default
res_both <- resShrunken
```

```{r res_bothgcsf}
contrast=c("treatment", "BOTH", "GCSF")
resUnshrunken <- results(
    dds,
    contrast = contrast,
    alpha = alpha)

# DESeqResults with shrunken log2 fold changes (LFC)
# help("lfcShrink", "DESeq2")
# Only `coef` or `contrast` can be specified, not both
# Use the correct `coef` number to modify from `resultsNames(dds)`
resShrunken <- lfcShrink(
    dds = dds,
    # coef = 2,
    contrast = contrast,
    res = resUnshrunken)

# Use shrunken LFC values by default
res_bothgcsf <- resShrunken
```

```{r res_bothza}
contrast=c("treatment", "BOTH", "ZA")
resUnshrunken <- results(
    dds,
    contrast = contrast,
    alpha = alpha)

# DESeqResults with shrunken log2 fold changes (LFC)
# help("lfcShrink", "DESeq2")
# Only `coef` or `contrast` can be specified, not both
# Use the correct `coef` number to modify from `resultsNames(dds)`
resShrunken <- lfcShrink(
    dds = dds,
    # coef = 2,
    contrast = contrast,
    res = resUnshrunken)

# Use shrunken LFC values by default
res_bothza <- resShrunken
```

We performed the analysis using a BH adjusted *P* value cutoff of `r alpha` and a log fold-change (LFC) ratio cutoff of `r lfc`.

## Plots 

### Volcano {.tabset}

A volcano plot compares significance (BH-adjusted *P* value) against fold change (log2) [@Cui:2003kh; @Li:2014fv]. Genes in the green box have an adjusted *P* value less than `r alpha` and an absolute log2Foldchange of at least `r lfc` and are likely to be the top candidate genes of interest. The top 20 genes by fold change are text labelled.

#### GCSF versus CTL

```{r plot_volcano_gcsf}
plotVolcano(res_gcsf, lfc = lfc,alpha = alpha, ntop=20, histograms=FALSE, gene2symbol=gene2symbol(bcb))
```

#### ZA versus CTL

```{r plot_volcano_za}
plotVolcano(res_za, lfc = lfc,alpha = alpha, ntop=20, histograms=FALSE, gene2symbol=gene2symbol(bcb))
```

#### BOTH versus CTL

```{r plot_volcano_both}
plotVolcano(res_both, lfc = lfc,alpha = alpha, ntop=20, histograms=FALSE, gene2symbol=gene2symbol(bcb))
```

#### BOTH versus GCSF

```{r plot_volcano_bothgcsf}
plotVolcano(res_bothgcsf, lfc = lfc,alpha = alpha, ntop=20, histograms=FALSE, gene2symbol=gene2symbol(bcb))
```

#### BOTH versus ZA

```{r plot_volcano_bothza}
plotVolcano(res_bothza, lfc = lfc,alpha = alpha, ntop=20, histograms=FALSE, gene2symbol=gene2symbol(bcb))
```


### Heatmap {.tabset}

This plot shows only differentially expressed genes on a per-sample basis. We have scaled the data by row and used the `ward.D2` method for clustering [@WardJr:1963eu]. 

One take away from these results is that the individual treatment samples cluster well within a treatment for any of the differentially expressed gene sets. Further, on a larger level, with the exception of genes that are differentially expressed between ZA and CTL samples, the samples cluster so that BOTH samples cluster with GCSF and ZA samples cluster with CTL for the differentially expressed genes.

The genes shown have an adjusted pvalue of less than `r alpha` and an absolute log2Foldchange of at least `r lfc`

#### GCSF versus CTL

```{r plot_deg_heatmap_gcsf}
plotDEGHeatmapJH(res_gcsf, mylfc=lfc, myalpha=alpha, bcb, rld_object = rld)

```

#### ZA versus CTL
- no pvalue cutoff
```{r plot_deg_heatmap_za}
plotDEGHeatmapJH(res_za, mylfc=lfc, myalpha=alpha, bcb, rld_object = rld)
```

#### BOTH versus CTL
```{r plot_deg_heatmap_both}
plotDEGHeatmapJH(res_both, mylfc=lfc, myalpha=alpha, bcb, rld_object = rld)
```

#### BOTH versus GCSF
```{r plot_deg_heatmap_bothgcsf}
plotDEGHeatmapJH(res_bothgcsf, mylfc=lfc, myalpha=alpha, bcb, rld_object = rld)
```

#### BOTH versus ZA
```{r plot_deg_heatmap_bothza}
plotDEGHeatmapJH(res_bothza, mylfc=lfc, myalpha=alpha, bcb, rld_object = rld)
```

### Top genes {.tabset}

These tables show the **top 20 DE genes**, as determined either by adjusted pvalue (when at least two replicates are available per sample class) or by log2FoldChange (when only one replicate is available for a sample class i.e. any of the ZA comparisons)

See the full Excel files of results to see all potentially DE genes.

#### GCSF versus CTL

```{r tabledeg_gcsf, results="asis"}
output_results(res_object=res_gcsf, bcb_object=bcb)
saveData(res_gcsf, dir="data/2018-03-30")
```

#### ZA versus CTL
- no pvalue cutoff
```{r tabledeg_za, results="asis"}
output_results(res_object=res_za, bcb_object=bcb, keep_stat=FALSE)
saveData(res_za, dir="data/2018-03-30")
```

#### BOTH versus CTL
```{r tabledeg_both, results="asis"}
output_results(res_object=res_both, bcb_object=bcb)
saveData(res_both, dir="data/2018-03-30")
```

#### BOTH versus GCSF
```{r tabledeg_bothgcsf, results="asis"}
output_results(res_object=res_bothgcsf, bcb_object=bcb)
saveData(res_bothgcsf, dir="data/2018-03-30")
```

#### BOTH versus ZA
```{r tabledeg_bothza, results="asis"}
output_results(res_object=res_bothza, bcb_object=bcb, keep_stat=FALSE)
saveData(res_bothza, dir="data/2018-03-30")
```


----

## Functional Analyses 

```{r setenv}
# set some perl system vars so the correct version of perl is used in the revigo.pl system calls
Sys.setenv(PATH="/usr/local/Cellar/perl/5.26.1/bin/:$PATH")
Sys.setenv(PERL5LIB="/usr/local/Cellar/perl/5.26.1/lib/perl5/site_perl/5.26.1/")
```


Using the top 200 most DE genes (i.e. the top 200 genes as ranked by ascending absolute log2FC) as input.

### GCSF versus CTL {.tabset}
```{r gprofrevigo_gcsf, results='asis'}
run_gprofiler_revigo(res_gcsf )
```

### ZA versus CTL {.tabset}
```{r gprofrevigo_za, results='asis'}
run_gprofiler_revigo(res_za )
```

### BOTH versus CTL {.tabset}
```{r gprofrevigo_both, results='asis'}
run_gprofiler_revigo(res_both )
```

### BOTH versus GCSF {.tabset}
```{r gprofrevigo_bothgcsf, results='asis'}
run_gprofiler_revigo(res_bothgcsf )
```

### BOTH versus ZA {.tabset}
```{r gprofrevigo_bothza, results='asis'}
run_gprofiler_revigo(res_bothza )
```


----


# File downloads

The results are saved as gzip-compressed comma separated values (CSV). Gzip compression is natively supported on [macOS][] and Linux-based operating systems. If you're running Windows, we recommend installing [7-Zip][]. CSV files can be opened in [Excel][] or [RStudio][].

## Count matrices

- [`normalizedCounts.csv.gz`](`r file.path(countsDir, "normalizedCounts.csv.gz")`): Use to evaluate individual genes and/or generate plots. These counts are normalized for the variation in sequencing depth across samples.
- [`tpm.csv.gz`](`r file.path(countsDir, "tpm.csv.gz")`): Transcripts per million, scaled by length and also suitable for plotting.
- [`rawCounts.csv.gz`](`r file.path(countsDir, "rawCounts.csv.gz")`): Only use to perform a new differential expression analysis. These counts will vary across samples due to differences in sequencing depth, and have not been normalized. Do not use this file for plotting genes.

## Differentially expressed genes (DEG)

All files can be found in this [folder](./results/differential_expression/). 

\*all.csv - statistics and annotations for all genes       

DEG tables are sorted by BH-adjusted P value, and contain the following columns:

- `ensgene`: [Ensembl][] gene identifier.
- `baseMean`: Mean of the normalized counts per gene for all samples.
- `log2FoldChange`: log2 fold change.
- `lfcSE`: log2 standard error.
- `stat`: Wald statistic.
- `pvalue`: Walt test *P* value.
- `padj`: BH adjusted Wald test *P* value (corrected for multiple comparisons; aka FDR).
- `externalGeneName`: [Ensembl][] name (a.k.a. symbol).
- `description`: [Ensembl][] description.
- `geneBiotype`: [Ensembl][] biotype (e.g. `protein_coding`).


```{r footer, child="_footer.Rmd", eval=file.exists("_footer.Rmd")}
```
