---
title: "Are the young knockouts more like old wildtype mice?"
author: "`r getOption('author')`"
date: "`r Sys.Date()`"
bibliography: bibliography.bib
output:
  html_document:
    toc: true
    toc_float:
      collapsed: false
      smooth_scroll: true
    number_sections: true
    theme: united
    hightlight: tango
      
params:
    bcb: "data/bcb_sub.rda"
    design: !r formula(~newgroup)
    contrast: !r contrast=c("newgroup", "put_old", "put_young")
    alpha: 0.1
    lfc: 0.5849625
    outputDir: "."
---


```{r setup, message=FALSE}
library(DESeq2)
library(bcbioRNASeq)

# Shared RMarkdown settings
prepareRNASeqTemplate()
if (file.exists("setup.R")) {
    source("setup.R")
}

# Load bcbioRNADataSet object
if (exists("bcb", inherits = FALSE)) {
    stop("bcb already exists in environment")
}
bcbName <- load(params$bcb)
bcb <- get(bcbName, inherits = FALSE)
if (!is(bcb, "bcbioRNADataSet")) {
    stop("bcb must bcbioRNADataSet class object")
}

# Design formula
# help("design", "DESeq2")
design <- params$design

# Contrast
# help("results", "DESeq2")
#   1. Design matrix parameter.
#   2. Numerator for LFC (expt).
#   3. Denominator for LFC (control).
contrast <- params$contrast

# Significance cutoffs
alpha <- params$alpha
lfc <- params$lfc

# Directory paths
outputDir <- params$outputDir
countsDir <- file.path(outputDir, "results", "counts")
deDir <- file.path(outputDir, "results", "differential_expression", "age_effects", "knockout")

source("functions.R")
```

```{r header, child="_header.Rmd", eval=file.exists("_header.Rmd")}
```

Here I grouped the young knockout mice with the old wildtype mice (KO sibs) and compared them to the young wildtype mice (KO sibs) to see if there are any differences. If the young knockouts are different from the young wildtype mice in the same way as the old wildtype mice for any genes, those genes should show up. That is to say, this analysis will find genes where the young knockout mice most resemble the old wildtype mice. To maintain balance between the two groups in the new putatitve old group (WT old and KO young samples), I dropped the FF_WT_old_2 sample.

```{r dds, results="hide"}
colData(bcb)$group <- factor(paste0(colData(bcb)$class, "-", colData(bcb)$age))

colData(bcb)$newgroup <- ifelse(colData(bcb)$group=="SCKO-young", "put_old", ifelse(colData(bcb)$group=="FF_WT-young", "put_young", ifelse(colData(bcb)$group=="FF_WT-old", "put_old", NA)))

bcb_sub <- bcb[,which(!is.na(colData(bcb)$newgroup) & !colData(bcb)$sampleid=="FF_WT_old_2"), skipNorm=TRUE]


dds <- DESeqDataSetFromTximport(
    txi = txi(bcb_sub),
    colData = colData(bcb_sub),
    design = formula(~newgroup)) %>%
    DESeq
rld <- rlog(dds)
```

# Alpha level (FDR) cutoffs {.tabset}

Let's take a look at the number of genes we get with different false discovery rate (FDR) cutoffs. These tests subset *P* values that have been multiple test corrected using the Benjamini Hochberg (BH) method [@Benjamini:1995ws].

This comparison accounts for any genotype based effects (i.e. these are the changes that occur with age that are not affected by genotype).

```{r alpha_summary, results="asis"}
alphaSummary(dds, contrast=c("newgroup", "put_young", "put_old"))
```

# Results

```{r res}
resUnshrunken <- results(
    dds,
    contrast = contrast,
    alpha = alpha)

# DESeqResults with shrunken log2 fold changes (LFC)
# help("lfcShrink", "DESeq2")
# Only `coef` or `contrast` can be specified, not both
# Use the correct `coef` number to modify from `resultsNames(dds)`
resShrunken <- lfcShrink(
    dds = dds,
    # coef = 2,
    contrast = contrast,
    res = resUnshrunken)

# Use shrunken LFC values by default
res <- resShrunken
```

We performed the analysis using a BH adjusted *P* value cutoff of `r alpha` and a log2 fold-change (LFC) ratio cutoff of `r lfc`.

# Plots

## Mean average (MA)

An MA plot compares transformed counts on `M` (log ratio) and `A` (mean average) scales [@Yang:2002ty]. Points in red are significantly differentially expressed.

```{r plot_ma}
plotMA(res)
```

## Volcano 

A volcano plot compares significance (BH-adjusted *P* value) against fold change (log2) [@Cui:2003kh; @Li:2014fv]. Teh green box covers genes differentially expressed (DE) with a BH adjusted *P* value cutoff of less than `r alpha` and an absolute log2 fold-change (LFC) ratio  of at least `r lfc`. The top 10 genes as assessed by absolute log2 fold change are labelled.

```{r plot_volcano}
plotVolcano(res, lfc = lfc, alpha = alpha, ntop=10, histograms=FALSE)
```


## Heatmap{.tabset}

This plot shows only differentially expressed genes on a per-sample basis. We have scaled the data by row and used the `ward.D2` method for clustering [@WardJr:1963eu].

```{r plot_deg_heatmap, fig.height=12}
plotDEGHeatmap(res, counts = rld)
```

## Top tables of differentially expressed genes

```{r results_tables, results="asis"}
resTbl <- resultsTables(res, lfc = lfc, dir = deDir)
```

Only the top up- and down-regulated genes (arranged by log2 fold change) are shown.

Tables are sorted by BH-adjusted P value, and contain the following columns:

- `ensgene`: [Ensembl][] gene identifier.
- `baseMean`: Mean of the normalized counts per gene for all samples.
- `log2FoldChange`: log2 fold change.
- `lfcSE`: log2 standard error.
- `stat`: Wald statistic.
- `pvalue`: Walt test *P* value.
- `padj`: BH adjusted Wald test *P* value (corrected for multiple comparisons; aka FDR).
- `externalGeneName`: [Ensembl][] name (a.k.a. symbol).
- `description`: [Ensembl][] description.
- `geneBiotype`: [Ensembl][] biotype (e.g. `protein_coding`).

```{r top_tables, results="asis"}
topTables(resTbl)
```

## Functional analysis

Using the significant DE genes (padj < `r alpha`) identified, a list of statistically enriched gene ontology (GO), [`Human Phenotype Ontology`][] (HP) terms, and KEGG pathways (keg) was generated using the program [gprofileR][] [@Reimand:2007].

The full list of enriched categories can be downloaded below.

A list including only the significant GO Biological Process (BP) terms was then used as input to [REViGO][] [@Supek:2011], which collapsed redundant and semantically-related terms and output the most significantly enriched functional categories.

```{r gprofiler_revigo, results='asis'}
run_gprofiler_revigo(my_results=res)
```

# File downloads 

The results are saved as gzip-compressed comma separated values (CSV). Gzip compression is natively supported on [macOS][] and Linux-based operating systems. If you're running Windows, we recommend installing [7-Zip][]. CSV files can be opened in [Excel][] or [RStudio][].

## Count matrices

- [`normalizedCounts.csv.gz`](`r file.path(countsDir, "normalizedCounts.csv.gz")`): Use to evaluate individual genes and/or generate plots. These counts are normalized for the variation in sequencing depth across samples.
- [`tpm.csv.gz`](`r file.path(countsDir, "tpm.csv.gz")`): Transcripts per million, scaled by length and also suitable for plotting.
- [`rawCounts.csv.gz`](`r file.path(countsDir, "rawCounts.csv.gz")`): Only use to perform a new differential expression analysis. These counts will vary across samples due to differences in sequencing depth, and have not been normalized. Do not use this file for plotting genes.

## Differentially expressed genes (DEG)

All files can be found in this [folder](./results/differential_expression/age_effects/transgenic). 

\*_all.csv.gz - statistics and annotations for all genes       
\*_deg.csv.gz - statistics and annotations for differentially expressed genes
\*_deg_lfc_down.csv.gz - statistics and annotations for differentially expressed genes that are up in first listed sample class (i.e. if filename is 1_vs_2, genes are more highly expressed in 1)
group_tg_young_vs_tg_old_deg_lfc_up.csv.gz - statistics and annotations for differentially expressed genes that are down in first listed sample class (i.e. if filename is 1_vs_2, genes are more highly expressed in 1)

DEG tables are sorted by BH-adjusted P value, and contain the following columns:

- `ensgene`: [Ensembl][] gene identifier.
- `baseMean`: Mean of the normalized counts per gene for all samples.
- `log2FoldChange`: log2 fold change.
- `lfcSE`: log2 standard error.
- `stat`: Wald statistic.
- `pvalue`: Walt test *P* value.
- `padj`: BH adjusted Wald test *P* value (corrected for multiple comparisons; aka FDR).
- `externalGeneName`: [Ensembl][] name (a.k.a. symbol).
- `description`: [Ensembl][] description.
- `geneBiotype`: [Ensembl][] biotype (e.g. `protein_coding`).

##Functional Enrichments

All files can be found in this [folder](./results/differential_expression/age_effects/transgenic/). 

gprofiler_results_\*.csv

Tables are sorted by pvalue and contain the followoing columns:

- `term.id`: Functional term id (domain dependent)
- `term.name`: Functional term name
- `domain`: Functional database  
    - `BP` - Gene Ontology Biological Process
    - `MF` - Gene Ontology Molecular Function   
    - `CC` - Gene Ontology Cellular Component  
    - `tf` - transcription factor motif enrichment  
    - `cor` - [CORUM][] protein complexes   
    - `hp` - [`Human Phenotype Ontology`][]  
    - `mi` - miRBase microRNA targets
- `p.value` - multiple test adjusted pvalue
- `term.size` - number of genes in the functional term
- `query.size` - number of differentially expressed genes examined
- `overlap.size` - number of differentially expressed genes in the functional term
- `assoc.gene.ids` - gene ids of differentially expressed genes in the functional term

[Corum]: http://mips.helmholtz-muenchen.de/corum/
[REViGO]: http://revigo.irb.hr/
[gprofileR]: http://biit.cs.ut.ee/gprofiler/
[`Human Phenotype Ontology`]: https://human-phenotype-ontology.github.io/



```{r footer, child="_footer.Rmd", eval=file.exists("_footer.Rmd")}
```