---
title: "Young versus Old Differential Expression"
author: "`r getOption('author')`"
date: "`r Sys.Date()`"
bibliography: bibliography.bib
params:
    bcb: "data/bcb_sub.rda"
    design: !r formula(~group)
    contrast: !r c("group", "TG-yng", "TG-old")
    alpha: 0.1
    lfc: 0.5849625
    outputDir: "."
---

```{r setup, message=FALSE}
library(DESeq2)
library(bcbioRNASeq)

# Shared RMarkdown settings
prepareRNASeqTemplate()
if (file.exists("setup.R")) {
    source("setup.R")
}

# Load bcbioRNADataSet object
if (exists("bcb", inherits = FALSE)) {
    stop("bcb already exists in environment")
}
bcbName <- load(params$bcb)
bcb <- get(bcbName, inherits = FALSE)
if (!is(bcb, "bcbioRNADataSet")) {
    stop("bcb must bcbioRNADataSet class object")
}

# Design formula
# help("design", "DESeq2")
design <- params$design

# Contrast
# help("results", "DESeq2")
#   1. Design matrix parameter.
#   2. Numerator for LFC (expt).
#   3. Denominator for LFC (control).
contrast <- params$contrast

# Significance cutoffs
alpha <- params$alpha
lfc <- params$lfc

# Directory paths
outputDir <- params$outputDir
countsDir <- file.path(outputDir, "results", "counts")
deDir <- file.path(outputDir, "results", "differential_expression", "age_effects")
```

```{r header, child="_header.Rmd", eval=file.exists("_header.Rmd")}
```

Here I looked at the effects of aging on gene expression, specific to the transgenic.

```{r dds, results="hide"}


colData(bcb)$group <- factor(paste0(colData(bcb)$class, "-", colData(bcb)$age))
design <- formula(~group)

dds <- DESeqDataSetFromTximport(
    txi = txi(bcb),
    colData = colData(bcb),
    design = design %>%
    DESeq)
rld <- rlog(dds)
```

# Alpha level (FDR) cutoffs {.tabset}

Let's take a look at the number of genes we get with different false discovery rate (FDR) cutoffs. These tests subset *P* values that have been multiple test corrected using the Benjamini Hochberg (BH) method [@Benjamini:1995ws].

This comparison accounts for any genotype based effects (i.e. these are the changes that occur with age that are not affected by genotype).

```{r alpha_summary, results="asis"}
alphaSummary(dds, contrast=c("age", "young", "old"))
```

# Results

```{r res}
resUnshrunken <- results(
    dds,
    contrast = contrast,
    alpha = alpha)

# DESeqResults with shrunken log2 fold changes (LFC)
# help("lfcShrink", "DESeq2")
# Only `coef` or `contrast` can be specified, not both
# Use the correct `coef` number to modify from `resultsNames(dds)`
resShrunken <- lfcShrink(
    dds = dds,
    # coef = 2,
    contrast = contrast,
    res = resUnshrunken)

# Use shrunken LFC values by default
res <- resShrunken
```

We performed the analysis using a BH adjusted *P* value cutoff of `r alpha` and a log2 fold-change (LFC) ratio cutoff of `r lfc`.

# Plots

## Mean average (MA)

An MA plot compares transformed counts on `M` (log ratio) and `A` (mean average) scales [@Yang:2002ty]. Points in red are significantly differentially expressed.

```{r plot_ma}
plotMA(res)
```

## Volcano 

A volcano plot compares significance (BH-adjusted *P* value) against fold change (log2) [@Cui:2003kh; @Li:2014fv]. Teh green box covers genes differentially expressed (DE) with a BH adjusted *P* value cutoff of less than `r alpha` and an absolute log2 fold-change (LFC) ratio  of at least `r lfc`. The top 10 genes as assessed by absolute log2 fold change are labelled.

```{r plot_volcano}
plotVolcano(res, lfc = lfc, alpha = alpha, ntop=10)
```


## Heatmap{.tabset}

This plot shows only differentially expressed genes on a per-sample basis. We have scaled the data by row and used the `ward.D2` method for clustering [@WardJr:1963eu].

```{r plot_deg_heatmap}
plotDEGHeatmap(res, counts = rld)
```

# File downloads 

The results are saved as gzip-compressed comma separated values (CSV). Gzip compression is natively supported on [macOS][] and Linux-based operating systems. If you're running Windows, we recommend installing [7-Zip][]. CSV files can be opened in [Excel][] or [RStudio][].

## Count matrices

- [`normalizedCounts.csv.gz`](`r file.path(countsDir, "normalizedCounts.csv.gz")`): Use to evaluate individual genes and/or generate plots. These counts are normalized for the variation in sequencing depth across samples.
- [`tpm.csv.gz`](`r file.path(countsDir, "tpm.csv.gz")`): Transcripts per million, scaled by length and also suitable for plotting.
- [`rawCounts.csv.gz`](`r file.path(countsDir, "rawCounts.csv.gz")`): Only use to perform a new differential expression analysis. These counts will vary across samples due to differences in sequencing depth, and have not been normalized. Do not use this file for plotting genes.

## Differentially expressed genes (DEG)

```{r results_tables, results="asis"}
resTbl <- resultsTables(res, lfc = lfc, dir = deDir)
```

All files can be found in this [folder](.results/differential_expression). 

age_young_vs_old_all.csv.gz - statistics and annotations for all genes       
age_young_vs_old_deg.csv.gz - statistics and annotations for differentially expressed genes
age_young_vs_old_deg_lfc_down.csv.gz - statistics and annotations for differentially expressed genes that are up in first listed sample class (i.e. if filename is 1_vs_2, genes are more highly expressed in 1)
age_young_vs_old_deg_lfc_up.csv.gz - statistics and annotations for differentially expressed genes that are down in first listed sample class (i.e. if filename is 1_vs_2, genes are more highly expressed in 1)

DEG tables are sorted by BH-adjusted P value, and contain the following columns:

- `ensgene`: [Ensembl][] gene identifier.
- `baseMean`: Mean of the normalized counts per gene for all samples.
- `log2FoldChange`: log2 fold change.
- `lfcSE`: log2 standard error.
- `stat`: Wald statistic.
- `pvalue`: Walt test *P* value.
- `padj`: BH adjusted Wald test *P* value (corrected for multiple comparisons; aka FDR).
- `externalGeneName`: [Ensembl][] name (a.k.a. symbol).
- `description`: [Ensembl][] description.
- `geneBiotype`: [Ensembl][] biotype (e.g. `protein_coding`).

# Top tables {.tabset}

Only the top up- and down-regulated genes (arranged by log2 fold change) are shown.

```{r top_tables, results="asis"}
topTables(resTbl)
```


```{r footer, child="_footer.Rmd", eval=file.exists("_footer.Rmd")}
```