---
title: "Young versus Old Differential Expression"
author: "`r getOption('author')`"
date: "`r Sys.Date()`"
bibliography: bibliography.bib
output:
  html_document:
    toc: true
    toc_float:
      collapsed: false
      smooth_scroll: true
    number_sections: true
    theme: united
    hightlight: tango
params:
    bcb: "data/bcb_sub.rda"
    design: !r formula(~class+age)
    contrast: !r c("age", "young", "old")
    alpha: 0.1
    lfc: 0.5849625
    outputDir: "."
---

```{r setup, message=FALSE}
library(DESeq2)
library(bcbioRNASeq)

# Shared RMarkdown settings
prepareRNASeqTemplate()
if (file.exists("setup.R")) {
    source("setup.R")
}

# Load bcbioRNADataSet object
if (exists("bcb", inherits = FALSE)) {
    stop("bcb already exists in environment")
}
bcbName <- load(params$bcb)
bcb <- get(bcbName, inherits = FALSE)
if (!is(bcb, "bcbioRNADataSet")) {
    stop("bcb must bcbioRNADataSet class object")
}

# Design formula
# help("design", "DESeq2")
design <- params$design

# Contrast
# help("results", "DESeq2")
#   1. Design matrix parameter.
#   2. Numerator for LFC (expt).
#   3. Denominator for LFC (control).
contrast <- params$contrast

# Significance cutoffs
alpha <- params$alpha
lfc <- params$lfc

# Directory paths
outputDir <- params$outputDir
countsDir <- file.path(outputDir, "results", "counts")
deDir <- file.path(outputDir, "results", "differential_expression", "age_effects", "general")

source("functions.R")
```

```{r header, child="_header.Rmd", eval=file.exists("_header.Rmd")}
```

Here I looked at the effects of aging on gene expression, while controlling for genotype.

```{r dds, results="hide"}
dds <- DESeqDataSetFromTximport(
    txi = txi(bcb),
    colData = colData(bcb),
    design = design) %>%
    DESeq
rld <- rlog(dds)
```

# Alpha level (FDR) cutoffs {.tabset}

Let's take a look at the number of genes we get with different false discovery rate (FDR) cutoffs. These tests subset *P* values that have been multiple test corrected using the Benjamini Hochberg (BH) method [@Benjamini:1995ws].

This comparison accounts for any genotype based effects (i.e. these are the changes that occur with age that are not affected by genotype).

```{r alpha_summary, results="asis"}
alphaSummary(dds, contrast=c("age", "young", "old"))
```

# Results

```{r res}
resUnshrunken <- results(
    dds,
    contrast = contrast,
    alpha = alpha)

# DESeqResults with shrunken log2 fold changes (LFC)
# help("lfcShrink", "DESeq2")
# Only `coef` or `contrast` can be specified, not both
# Use the correct `coef` number to modify from `resultsNames(dds)`
resShrunken <- lfcShrink(
    dds = dds,
    # coef = 2,
    contrast = contrast,
    res = resUnshrunken)

# Use shrunken LFC values by default
res <- resShrunken
```

We performed the analysis using a BH adjusted *P* value cutoff of `r alpha` and a log2 fold-change (LFC) ratio cutoff of `r lfc`.

## Plots

### Mean average (MA)

An MA plot compares transformed counts on `M` (log ratio) and `A` (mean average) scales [@Yang:2002ty]. Points in red are significantly differentially expressed.

```{r plot_ma}
plotMA(res)
```

### Volcano 

A volcano plot compares significance (BH-adjusted *P* value) against fold change (log2) [@Cui:2003kh; @Li:2014fv]. Teh green box covers genes differentially expressed (DE) with a BH adjusted *P* value cutoff of less than `r alpha` and an absolute log2 fold-change (LFC) ratio  of at least `r lfc`. The top 10 genes as assessed by absolute log2 fold change are labelled.

```{r plot_volcano}
plotVolcano(res, lfc = lfc, alpha = alpha, ntop=10, histograms=FALSE)
```


### Heatmap{.tabset}

This plot shows only differentially expressed genes on a per-sample basis. We have scaled the data by row and used the `ward.D2` method for clustering [@WardJr:1963eu].

```{r plot_deg_heatmap}
plotDEGHeatmap(res, counts = rld)
```

## Top tables of Differentially Expressed Genes {.tabset}

Only the top up- and down-regulated genes (arranged by log2 fold change) are shown.

```{r top_tables, results="asis"}
topTables(resTbl)
```

## Functional analysis

Using the significant DE genes (padj < `r alpha`) identified, a list of statistically enriched gene ontology (GO), [`Human Phenotype Ontology`][] (HP) terms, and KEGG pathways (keg) was generated using the program [gprofileR][] [@Reimand:2007].

A list including only the significant GO Biological Process (BP) terms was then used as input to [REViGO][] [@Supek:2011], which collapsed redundant and semantically-related terms and output the most significantly enriched functional categories.

```{r functional_analysis_function, echo=FALSE}
library(gProfileR)
library(treemap)
library(RCurl)
library(dplyr)
library(stringr)

myresContrastName <- function(res) {
  mcols(res)[2L, 2L] %>%
    str_replace("^.*:\\s", "") %>%
    str_replace_all("_", " ") %>%
    # Improve appearance for difference of differences
    str_replace_all("\\+", " \\+\n    ")
}

run_gprofiler_revigo <- function(my_results, 
                                 num_sig=200, 
                                 gp_org="mmusculus", 
                                 gp_ordered=FALSE, 
                                 gp_correction="fdr", 
                                 rev_cutoff=0.7, 
                                 rev_organism="Mus musculus")  {
  if( !file.exists("revigo.pl")){
    message("revigo.pl needs to be in the working directory")
  }
  
  contrast <- myresContrastName(my_results)
  fileStem <- snake(contrast)
  
  # get significant genes
  sig_genes <-     my_results %>% tbl_df() %>% 
    rownames_to_column(var="ensemblId") %>% 
    arrange(padj) %>% 
    select(ensemblId) %>% 
    slice(., 1:num_sig) %>% 
    unlist() %>% 
    as.vector()
  
  #run gprofileR
  gprofiler_results <- gprofiler(query = sig_genes, 
                                 organism = gp_org, 
                                 ordered_query = gp_ordered, 
                                 exclude_iea = F, 
                                 max_set_size = 0, 
                                 correction_method = "fdr", 
                                 hier_filtering = "none", 
                                 domain_size = "annotated", 
                                 custom_bg = "")
  
  allterms <- gprofiler_results$term.id
  GOs <- allterms[grep("GO:", allterms)]
  pvals <- gprofiler_results$p.value[grep("GO:", allterms)]
  
  gprofiler_results.simplified <- gprofiler_results[, c("term.id", "term.name", "domain", "p.value", "term.size", "query.size", "overlap.size", "intersection")]
  names(gprofiler_results.simplified) <- c("term.id", "term.name",  "domain", "p.value","term.size", "query.size", "overlap.size", "assoc.gene.ids")
  
  # output to file
  write.csv(gprofiler_results.simplified, file.path(deDir, paste("gprofiler_results_", fileStem, ".csv", sep="" )))
  
  # Revigo 
  ## Parameters to change
  cutoff <- rev_cutoff #Allowed values: "0.90" "0.70" "0.50" "0.40" 
  organism <- "Mus musculus" #Allowed values: See organism.list below
  isPValue <- "yes" #Allowed values: "yes"  "no"
  whatIsBetter <- "higher" #Allowed values: "higher" "lower" "absolute" "abs_log"
  measure <- "SIMREL" #Allowed values: "RESNIK" "LIN" "SIMREL" "JIANG"
  domain <- "BP"
  
  #Do not change below
  organism.list <- list(
    "whole UniProt"=0, 
    "Homo sapiens"=9606,
    "Mus musculus"=10090,
    "Rattus norvegicus"=10116,
    "Bos taurus"=9913,
    "Gallus gallus"=9031,
    "Danio rerio"=7955,
    "Takifugu rubripes"=31033,
    "Xenopus laevis"=8355,
    "Drosophila melanogaster"=7227,
    "Caenorhabditis elegans"=6239,
    "Arabidopsis thaliana"=3702,
    "Oryza sativa"=39947,
    "Zea mays"=4577,
    "Saccharomyces cerevisiae"=4932,
    "Schizosaccharomyces pombe"=4896,
    "Dictyostelium discoideum"=44689,
    "Plasmodium falciparum"=5833,
    "Chlamydomonas reinhardtii"=3055,
    "Escherichia coli"=83333,
    "Bacillus subtilis"=1423,
    "Pseudomonas aeruginosa"=287,
    "Mycobacterium tuberculosis"=1773,
    "Mycoplasma genitalium"=2097,
    "Synechocystis sp."=1148
  )
  organism.db <- as.character(organism.list[organism])
  
  cat("**Biological Processes Domain**")
  
  domain="BP"
  mycommand=paste('revigo.pl -goterms', paste(GOs,collapse=","), '-gopvals', paste(pvals,collapse=","), '-cutoff', cutoff,  '-organism', organism.db, '-ispvalue', isPValue, '-whatisbetter', whatIsBetter, '-measure', measure, '-domain', domain,sep=" ")
  mytempfile <- tempfile()
  system2(command='perl', args=mycommand, stdout=mytempfile)
  source(mytempfile)
  
  cat("**Molecular Functions Domain**")
  
  domain="MF"
  mycommand=paste('revigo.pl -goterms', paste(GOs,collapse=","), '-gopvals', paste(pvals,collapse=","), '-cutoff', cutoff,  '-organism', organism.db, '-ispvalue', isPValue, '-whatisbetter', whatIsBetter, '-measure', measure, '-domain', domain,sep=" ")
  mytempfile <- tempfile()
  system2(command='perl', args=mycommand, stdout=mytempfile)
  source(mytempfile)
  
  cat("**Cellular Component Domain**")
  
  domain="CC"
  mycommand=paste('revigo.pl -goterms', paste(GOs,collapse=","), '-gopvals', paste(pvals,collapse=","), '-cutoff', cutoff,  '-organism', organism.db, '-ispvalue', isPValue, '-whatisbetter', whatIsBetter, '-measure', measure, '-domain', domain,sep=" ")
  mytempfile <- tempfile()
  system2(command='perl', args=mycommand, stdout=mytempfile)
  source(mytempfile)
}  
```

```{r gprofiler_revigo, results='asis'}
run_gprofiler_revigo(my_results=res)
```

# File downloads 

The results are saved as gzip-compressed comma separated values (CSV). Gzip compression is natively supported on [macOS][] and Linux-based operating systems. If you're running Windows, we recommend installing [7-Zip][]. CSV files can be opened in [Excel][] or [RStudio][].

## Count matrices

- [`normalizedCounts.csv.gz`](`r file.path(countsDir, "normalizedCounts.csv.gz")`): Use to evaluate individual genes and/or generate plots. These counts are normalized for the variation in sequencing depth across samples.
- [`tpm.csv.gz`](`r file.path(countsDir, "tpm.csv.gz")`): Transcripts per million, scaled by length and also suitable for plotting.
- [`rawCounts.csv.gz`](`r file.path(countsDir, "rawCounts.csv.gz")`): Only use to perform a new differential expression analysis. These counts will vary across samples due to differences in sequencing depth, and have not been normalized. Do not use this file for plotting genes.

## Differentially expressed genes (DEG)

```{r results_tables, results="asis"}
resTbl <- resultsTables(res, lfc = lfc, dir = deDir)
```

All files can be found in this [folder](./results/differential_expression/age_effects/general/). 

age_young_vs_old_all.csv.gz - statistics and annotations for all genes       
age_young_vs_old_deg.csv.gz - statistics and annotations for differentially expressed genes
age_young_vs_old_deg_lfc_down.csv.gz - statistics and annotations for differentially expressed genes that are up in first listed sample class (i.e. if filename is 1_vs_2, genes are more highly expressed in 1)
age_young_vs_old_deg_lfc_up.csv.gz - statistics and annotations for differentially expressed genes that are down in first listed sample class (i.e. if filename is 1_vs_2, genes are more highly expressed in 1)

DEG tables are sorted by BH-adjusted P value, and contain the following columns:

- `ensgene`: [Ensembl][] gene identifier.
- `baseMean`: Mean of the normalized counts per gene for all samples.
- `log2FoldChange`: log2 fold change.
- `lfcSE`: log2 standard error.
- `stat`: Wald statistic.
- `pvalue`: Walt test *P* value.
- `padj`: BH adjusted Wald test *P* value (corrected for multiple comparisons; aka FDR).
- `externalGeneName`: [Ensembl][] name (a.k.a. symbol).
- `description`: [Ensembl][] description.
- `geneBiotype`: [Ensembl][] biotype (e.g. `protein_coding`).

##Functional Enrichments

All files can be found in this [folder](./results/differential_expression/age_effects/general/). 

gprofiler_results_age_young_vs_old.csv

Tables are sorted by pvalue and contain the followoing columns:

- `term.id`: Functional term id (domain dependenet)
- `term.name`: Functional term name
- `domain`: Functional database  
    - `BP` - Gene Ontology Biological Process
    - `MF` - Gene Ontology Molecular Function   
    - `CC` - Gene Ontology Cellular Component  
    - `tf` - transcription factor motif enrichment  
    - `cor` - [CORUM][] protein complexes   
    - `hp` - [`Human Phenotype Ontology`][]  
    - `mi` - miRBase microRNA targets
- `p.value` - multiple test adjusted pvalue
- `term.size` - number of genes in the functional term
- `query.size` - number of differentially expressed genes examined
- `overlap.size` - number of differentially expressed genes in the functional term
- `assoc.gene.ids` - gene ids of differentially expressed genes in the functional term

[Corum]: http://mips.helmholtz-muenchen.de/corum/
[REViGO]: http://revigo.irb.hr/
[gprofileR]: http://biit.cs.ut.ee/gprofiler/
[`Human Phenotype Ontology`]: https://human-phenotype-ontology.github.io/

```{r footer, child="_footer.Rmd", eval=file.exists("_footer.Rmd")}
```