---
title: "Transgenic vs Knockout Differential Expression - Old samples"
author: "`r getOption('author')`"
date: "`r Sys.Date()`"
bibliography: bibliography.bib
params:
    bcb: "data/bcb_old.rda"
    design: !r formula(~sibset+generalGeno)
    contrast: !r c("generalGeno", "TG", "KO")
    alpha: 0.1
    lfc: 0.5849625
    outputDir: "."
---

```{r setup, message=FALSE}
library(DESeq2)
library(bcbioRNASeq)

# Shared RMarkdown settings
prepareRNASeqTemplate()
if (file.exists("setup.R")) {
    source("setup.R")
}

# Load bcbioRNADataSet object
if (exists("bcb", inherits = FALSE)) {
    stop("bcb already exists in environment")
}
bcbName <- load(params$bcb)
bcb <- get(bcbName, inherits = FALSE)
if (!is(bcb, "bcbioRNADataSet")) {
    stop("bcb must bcbioRNADataSet class object")
}

# Design formula
# help("design", "DESeq2")
design <- params$design

# Contrast
# help("results", "DESeq2")
#   1. Design matrix parameter.
#   2. Numerator for LFC (expt).
#   3. Denominator for LFC (control).
contrast <- params$contrast

# Significance cutoffs
alpha <- params$alpha
lfc <- params$lfc

# Directory paths
outputDir <- params$outputDir
countsDir <- file.path(outputDir, "results", "counts")
deDir <- file.path(outputDir, "results", "differential_expression", "KOvsTG", "old")
```

```{r header, child="_header.Rmd", eval=file.exists("_header.Rmd")}
```

To get a better idea of what's going on, here I looked at all the pairwise comparisons, Wildtype versus Transgenic, Knockout versus Wildtype and Knockout versus Transgenic. All the comparisons are adjusted for the sibling relationships between the samples (KO and FF_WT, TG and WT). 

```{r dds, results="hide"}
dds <- DESeqDataSetFromTximport(
    txi = txi(bcb),
    colData = colData(bcb),
    design = design) %>%
    DESeq
rld <- rlog(dds)
```

# Alpha level (FDR) cutoffs {.tabset}

Let's take a look at the number of genes we get with different false discovery rate (FDR) cutoffs. These tests subset *P* values that have been multiple test corrected using the Benjamini Hochberg (BH) method [@Benjamini:1995ws].

All of these comparisons account for any sibship batch effects.

## WT versus KO

```{r alpha_summary1, results="asis"}
alphaSummary(dds, contrast=c("generalGeno", "WT", "KO"))
```

## WT versus TG

```{r alpha_summary2, results="asis"}
alphaSummary(dds, contrast=c("generalGeno", "WT", "TG"))
```

## KO versus TG

```{r alpha_summary3, results="asis"}
alphaSummary(dds, contrast=c("generalGeno", "KO", "TG"))
```

# Results {.tabset}

## WT versus KO

```{r resWTKO}
contrast=c("generalGeno", "WT", "KO")

resUnshrunken <- results(
    dds,
    contrast = contrast,
    alpha = alpha)

# DESeqResults with shrunken log2 fold changes (LFC)
# help("lfcShrink", "DESeq2")
# Only `coef` or `contrast` can be specified, not both
# Use the correct `coef` number to modify from `resultsNames(dds)`
resShrunken <- lfcShrink(
    dds = dds,
    # coef = 2,
    contrast = contrast,
    res = resUnshrunken)

# Use shrunken LFC values by default
res_WTKO <- resShrunken
```

We performed the analysis using a BH adjusted *P* value cutoff of `r alpha` and a log2 fold-change (LFC) ratio cutoff of `r lfc`.

## TG versus WT

```{r resTGWT}
contrast=c("generalGeno", "TG", "WT")

resUnshrunken <- results(
    dds,
    contrast = contrast,
    alpha = alpha)

# DESeqResults with shrunken log2 fold changes (LFC)
# help("lfcShrink", "DESeq2")
# Only `coef` or `contrast` can be specified, not both
# Use the correct `coef` number to modify from `resultsNames(dds)`
resShrunken <- lfcShrink(
    dds = dds,
    # coef = 2,
    contrast = contrast,
    res = resUnshrunken)

# Use shrunken LFC values by default
res_TGWT <- resShrunken
```

We performed the analysis using a BH adjusted *P* value cutoff of `r alpha` and a log fold-change (LFC) ratio cutoff of `r lfc`.

## TG versus KO

```{r resTGKO}
contrast=c("generalGeno", "TG", "KO")

resUnshrunken <- results(
    dds,
    contrast = contrast,
    alpha = alpha)

# DESeqResults with shrunken log2 fold changes (LFC)
# help("lfcShrink", "DESeq2")
# Only `coef` or `contrast` can be specified, not both
# Use the correct `coef` number to modify from `resultsNames(dds)`
resShrunken <- lfcShrink(
    dds = dds,
    # coef = 2,
    contrast = contrast,
    res = resUnshrunken)

# Use shrunken LFC values by default
res_TGKO <- resShrunken
```

We performed the analysis using a BH adjusted *P* value cutoff of `r alpha` and a log fold-change (LFC) ratio cutoff of `r lfc`.

# Plots

## Mean average (MA) {.tabset}

An MA plot compares transformed counts on `M` (log ratio) and `A` (mean average) scales [@Yang:2002ty]. Points in red are significantly differentially expressed.

### WT versus KO

```{r plot_maWTKO}
plotMA(res_WTKO)
```

### TG versus WT

```{r plot_maTGWT}
plotMA(res_TGWT)
```

### TG versus KO 

```{r plot_maTGKO}
plotMA(res_TGKO)
```

## Volcano {.tabset}

A volcano plot compares significance (BH-adjusted *P* value) against fold change (log2) [@Cui:2003kh; @Li:2014fv]. Teh green box covers genes differentially expressed (DE) with a BH adjusted *P* value cutoff of less than `r alpha` and an absolute log2 fold-change (LFC) ratio  of at least `r lfc`. The top 10 genes as assessed by absolute log2 fold change are labelled.

### WT versus KO

```{r plot_volcanoWTKO}
plotVolcano(res_WTKO, lfc = lfc, alpha = alpha, ntop=10, histograms=FALSE)
```

### TG versus WT

```{r plot_volcanoTGWT}
plotVolcano(res_TGWT, lfc = lfc, alpha = alpha, ntop=10, histograms=FALSE)
```

### TG versus KO 

```{r plot_volcanoTGKO}
plotVolcano(res_TGKO, lfc = lfc, alpha = alpha, ntop=10, histograms=FALSE)
```

## Heatmap{.tabset}

This plot shows only differentially expressed genes on a per-sample basis. We have scaled the data by row and used the `ward.D2` method for clustering [@WardJr:1963eu].

### WT versus KO

```{r plot_deg_heatmapWTKO}
plotDEGHeatmap(res_WTKO, counts = rld)
```

### TG versus WT

```{r plot_deg_heatmapTGWT}
plotDEGHeatmap(res_TGWT, counts = rld)
```

### TG versus KO 

```{r plot_deg_heatmapTGKO}
plotDEGHeatmap(res_TGKO, counts = rld)
```


# File downloads 

The results are saved as gzip-compressed comma separated values (CSV). Gzip compression is natively supported on [macOS][] and Linux-based operating systems. If you're running Windows, we recommend installing [7-Zip][]. CSV files can be opened in [Excel][] or [RStudio][].



## Count matrices

- [`normalizedCounts.csv.gz`](`r file.path(countsDir, "normalizedCounts.csv.gz")`): Use to evaluate individual genes and/or generate plots. These counts are normalized for the variation in sequencing depth across samples.
- [`tpm.csv.gz`](`r file.path(countsDir, "tpm.csv.gz")`): Transcripts per million, scaled by length and also suitable for plotting.
- [`rawCounts.csv.gz`](`r file.path(countsDir, "rawCounts.csv.gz")`): Only use to perform a new differential expression analysis. These counts will vary across samples due to differences in sequencing depth, and have not been normalized. Do not use this file for plotting genes.

## Differentially expressed genes (DEG)

```{r results_tables, results="asis"}
resTblWTKO <- resultsTables(res_WTKO, lfc = lfc, dir = deDir)
resTblTGWT <- resultsTables(res_TGWT, lfc = lfc, dir = deDir)
resTblTGKO <- resultsTables(res_TGKO, lfc = lfc, dir = deDir)
```

All files can be found in this [folder](.results/differential_expression/KOvsTG/old). 

\*all.csv.gz - statistics and annotations for all genes       
\*deg.csv.gz - statistics and annotations for differentially expressed genes
\*deg_lfc_down.csv.gz - statistics and annotations for differentially expressed genes that are up in first listed sample class (i.e. if filename is 1_vs_2, genes are more highly expressed in 1)
\*deg_lfc_up.csv.gz - statistics and annotations for differentially expressed genes that are down in first listed sample class (i.e. if filename is 1_vs_2, genes are more highly expressed in 1)

DEG tables are sorted by BH-adjusted P value, and contain the following columns:

- `ensgene`: [Ensembl][] gene identifier.
- `baseMean`: Mean of the normalized counts per gene for all samples.
- `log2FoldChange`: log2 fold change.
- `lfcSE`: log2 standard error.
- `stat`: Wald statistic.
- `pvalue`: Walt test *P* value.
- `padj`: BH adjusted Wald test *P* value (corrected for multiple comparisons; aka FDR).
- `externalGeneName`: [Ensembl][] name (a.k.a. symbol).
- `description`: [Ensembl][] description.
- `geneBiotype`: [Ensembl][] biotype (e.g. `protein_coding`).

# Top tables {.tabset}

Only the top up- and down-regulated genes (arranged by log2 fold change) are shown.

## WT versus KO

```{r top_tablesWTKO, results="asis"}
topTables(resTblWTKO)
```

## TG versus WT

```{r top_tablesTGWT, results="asis"}
topTables(resTblTGWT)
```

## TG versus KO 

```{r top_tablesTGKO, results="asis"}
topTables(resTblTGKO)
```


```{r footer, child="_footer.Rmd", eval=file.exists("_footer.Rmd")}
```


