---
title: "Perrimon Lab blood scRNA-Seq quality control"
date: "`r BiocStyle::doc_date()`"
author: "Michael J. Steinbaugh"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(cache = TRUE,
                      cache.lazy = FALSE,
                      message = FALSE,
                      warning = FALSE)

options(readr.num_columns = 0)

library(biomaRt)
library(cowplot)
library(Matrix)
library(mltools)
library(pheatmap)
library(scales)
library(Seurat)
library(tools)

library(basejump)
library(tidyverse)
```

[BPF]: https://genome.med.harvard.edu
[Ensembl]: http://useast.ensembl.org/Drosophila_melanogaster/Info/Index
[FlyBase]: http://flybase.org
[inDrop]: http://1cell-bio.com
[Orchestra]: https://wiki.med.harvard.edu/Orchestra
[scRNA-Seq]: http://bcbio-nextgen.readthedocs.io/en/latest/contents/configuration.html#single-cell-rna-sequencing

[`bcbio-nextgen`]: https://bcbio-nextgen.readthedocs.io
[`bcl2fastq`]: https://support.illumina.com/downloads/bcl2fastq-conversion-software-v217.html
[`biomaRt`]: https://bioconductor.org/packages/release/bioc/html/biomaRt.html
[`rapmap`]: https://github.com/COMBINE-lab/RapMap
[`umis`]: https://github.com/vals/umis

# Overview

2000 cells were sequenced using the [inDrop][] v3 platform. 

- `M`: Data for another group
- `sorted`: Fluorescence-activated cell sorting (FACS)-sorted GFP+ immune cells.
- `unsorted`: All cells. Note that this group is also GFP+.

Only `sorted` and `unsorted` samples are relevant to the analysis. We'll filter `M` from further downstream quality control checks.


## [inDrop][] pipeline history

- v1: Original design
    - R1: metadata read
    - R2: biological read
- v2: Inversion of v1
- v3: Summer 2016 redesign, requiring manual demultiplexing
    - R1: biological read
    - R2: first half of the gel barcode
    - R3: library index
    - R4: second half of the gel barcode, the UMI, and a fraction of the polyA tail



# Analysis workflow

## Demultiplexing

We demultiplexed the raw Illumina BCL files provided by the [Harvard Biopolymers Facility][bpf] into FASTQ format using [`bcl2fastq`][]. This step generates files with the following suffixes:

- `_R1`: 61 bp read 1, transcript
- `_R2`: 8 bp index read 1 (i7), single cell barcode
- `_R3`: 8 bp index read 2 (i5), library index
- `_R4`: 14 bp read 2, barcode and unique molecular identifiers (UMIs)

We ran this on [Orchestra][] with the following settings:

```{bash, echo=TRUE, eval=FALSE}
module load seq/bcl2fastq/2.17.1.14
bcl2fastq --use-bases-mask y*,y*,y*,y* --mask-short-adapter-reads 0 --minimum-trimmed-read-length 0
```


## Count filtering and alignment

### [FlyBase][] transcriptome annotations

We aligned the counts using the latest transcriptome FASTA from [FlyBase][], [`dmel-all-transcript-r6.14.fasta`](ftp://ftp.flybase.net/genomes/Drosophila_melanogaster/dmel_r6.14_FB2017_01/fasta/dmel-all-transcript-r6.14.fasta.gz) (Feb 2017).

```{r bcbio, include=FALSE}
# HPC mount point over `sshfs` connection
hpc_dir <- file.path("~",
                     "Orchestra",
                     "bcbio",
                     "PIs",
                     "norbert_perrimon",
                     "blood_scrnaseq")
if (!length(dir(hpc_dir))) {
    stop("Project on HPC server not loaded.")
}
dir(hpc_dir)

# `bcbio-nextgen`
final_dir <- file.path(hpc_dir, "indrop_rnaseq", "final")
print(final_dir)
dir(final_dir)

# Default naming scheme is `indrop_rnaseq/final/YYYY-MM-DD_indrop_rnaseq`
summary_dir <- file.path(final_dir) %>%
    dir(full.names = TRUE) %>%
    .[grepl("/\\d{4}-\\d{2}-\\d{2}_indrop_rnaseq$", .)]
print(summary_dir)
dir(summary_dir)

# Create directories if necessary
if (!file.exists("data")) {
    dir.create("data")
}
if (!file.exists("results")) {
    dir.create("results")
}
```

### Additional [Ensembl][] annotations

We used [`biomaRt`][] to obtain additional metadata from [Ensembl][] for each transcript, namely the `biotype`.

```{r annotations}
mart <- biomaRt::useMart("ENSEMBL_MART_ENSEMBL", "dmelanogaster_gene_ensembl")
mart_attributes <- biomaRt::listAttributes(mart)
annotations <- biomaRt::getBM(mart = mart,
                              attributes = c("ensembl_gene_id",
                                             "ensembl_transcript_id",
                                             "external_gene_name",
                                             "external_transcript_name",
                                             "gene_biotype",
                                             "transcript_biotype",
                                             "chromosome_name"))
# Broad class definitions
coding <- c("protein_coding")
decaying <- c("non_stop_decay",
              "nonsense_mediated_decay")
noncoding <- c("known_ncrna",
               "lincRNA",
               "non_coding")
srna <- c("miRNA",
          "misc_RNA",
          "ribozyme",
          "rRNA",
          "scaRNA",
          "scRNA",
          "snoRNA",
          "snRNA",
          "sRNA")
annotations$broad_class <-
    dplyr::case_when(grepl("mito", annotations$chromosome_name) ~ "mito",
                     grepl("pseudo", annotations$gene_biotype) ~ "pseudo",
                     grepl("TR_", annotations$gene_biotype) ~ "TCR",
                     grepl("IG_", annotations$gene_biotype) ~ "IG",
                     annotations$gene_biotype %in% srna ~ "small",
                     annotations$gene_biotype %in% decaying ~ "decaying",
                     annotations$gene_biotype %in% noncoding ~ "noncoding",
                     annotations$gene_biotype %in% coding ~ "coding",
                     TRUE ~ "other")

coding <- annotations %>%
    subset(broad_class == "coding") %>%
    .$ensembl_transcript_id %>%
    basejump::sortUnique(.)
mito <- annotations %>%
    subset(broad_class == "mito") %>%
    .$ensembl_transcript_id %>%
    basejump::sortUnique(.)

save(annotations, file = "data/annotations.rda")
write.csv(annotations, "results/annotations.csv")
rm(mart, mart_attributes)
```

```{r tx2name}
tx2name <- annotations %>%
    .[, c("ensembl_transcript_id",
          "external_transcript_name")] %>%
    unique
save(tx2name, file = "data/tx2name.rda")
```

```{r matrix}
counts <- file.path(summary_dir, "tagcounts.mtx") %>%
    Matrix::readMM(.)
rownames(counts) <- file.path(summary_dir, "tagcounts.mtx.rownames") %>%
    readr::read_lines(.)
colnames(counts) <- file.path(summary_dir, "tagcounts.mtx.colnames") %>%
    readr::read_lines(.)
save(counts, file = "data/counts.rda")
```

```{r matrix_check, eval=FALSE, include=FALSE}
# Check the `colnames` and `rownames`
head(counts@Dimnames[[1]])
head(counts@Dimnames[[2]])
```

### Sample barcodes

We require there to be at least 10,000 reads to match a cellular barcode before aligning.

```{r sample_barcodes}
sample_barcodes <- file.path(hpc_dir, "sample_barcodes.log") %>%
    read.table(col.names = c("counts", "reverse_complement"))
metadata <- file.path("data-raw", "metadata.xlsx") %>%
    readxl::read_excel(.) %>%
    dplyr::left_join(sample_barcodes, by = "reverse_complement") %>%
    dplyr::mutate(unique = paste(run, reverse_complement, sep = "-"))
save(metadata, file = "data/metadata.rda")
rm(sample_barcodes)
printTable(metadata, caption = "Sample metadata")
```



# Quality control metrics

## Barcode plot

These are counts of how many reads are assigned to a given cellular barcode. The purpose of these histograms is that we should see one single large peak that represents cells that were encapsulated. If we see a strong shoulder, or a bimodal distribution of the cells, that can indicate a couple problems. It might be that there is free floating RNA, which happens when cells are dying. It could also be that there are a set of cells that failed for some reason. Finally, it could also be that there are biologically different types of cells, and one type is much smaller than the other. If this is the case we would expect to see less RNA being sequenced from the smaller cells.

Both the `sorted` and `unsorted` groups have an apparent bimodal distribution, which could be indicative of a large number of dying cells or low complexity cells, possibly red blood cells. We're not entirely sure what's behind this effect, and we are working on figuring this out with the Harvard Single Cell Core.

```{r barcode_plot}
read_barcode_file = function(filename) {
    return(read_tsv(filename, col_names = c("barcode", "count"), progress = FALSE))
}
barcode_plot = function(filename, sample) {
    bcs = read_barcode_file(filename)
    bcs_hist = hist(log10(bcs$count), plot = FALSE, n = 50)
    fLog = bcs_hist$count
    xLog = bcs_hist$mids
    y = fLog * (10^xLog)/sum(fLog * (10^xLog))
    p = qplot(10^xLog, y) +
        geom_point() +
        geom_line() +
        ggtitle(sample) +
        scale_x_log10(breaks = scales::trans_breaks("log10", function(x) 10^x),
                      labels = scales::trans_format("log10", scales::math_format(10^.x))) +
        xlab("number of reads assigned to a cell") +
        ylab("proportion of cells")
    print(p)
}
bcfiles <- list.files(hpc_dir, pattern = "*-barcodes.tsv",
                      recursive = TRUE,
                      include.dirs = TRUE,
                      full.names = TRUE)
names(bcfiles) <- gsub("-barcodes\\.tsv$", "", basename(bcfiles))
for (a in 1:length(bcfiles)) {
    barcode_plot(bcfiles[a], names(bcfiles)[a])
}
save(bcfiles, file = "data/bcfiles.rda")
```

```{r metrics}
# `rmarkdown::render()` doesn't appear to handle `dgTMatrix` objects properly.
# Fails on `colSums(counts)` unless we convert `counts` to a matrix first.
counts_matrix <- as.matrix(counts)
metrics <- data.frame(identifier = colnames(counts_matrix),
                      total_counts = colSums(counts_matrix),
                      genes_detected = colSums(counts_matrix > 0),
                      coding_counts = colSums(counts_matrix[rownames(counts_matrix) %in% coding, ]),
                      mito_counts = colSums(counts_matrix[rownames(counts_matrix) %in% mito, ])) %>%
    tidyr::separate(identifier,
                    c("sample_barcode", "cellular_barcode"),
                    sep = ":",
                    remove = FALSE) %>%
    dplyr::mutate(log_detected_per_count = log(genes_detected) / log(total_counts),
                  percent_mito = mito_counts / total_counts,
                  sample_barcode = gsub("^.*-([ACGT]+)$", "\\1", sample_barcode)) %>%
    dplyr::left_join(metadata, by = c("sample_barcode" = "reverse_complement")) %>%
    dplyr::select(sample, dplyr::everything()) %>%
    dplyr::arrange(sample, identifier)
save(metrics, file = "data/metrics.rda")
write.csv(metrics, "results/metrics.csv")
rm(counts_matrix)
```

We have filtered the `M` samples from subsequent quality control analysis.

```{r metrics_filter}
metrics <- dplyr::filter(metrics, sample != "M")
```


## Total number of cells

More cells were analyzed in the FACS sorted group than the expected amount of 2,000. We obtained a low cell count in the unsorted group.

```{r total_cells}
total_cells <- metrics %>%
    dplyr::group_by(sample) %>%
    dplyr::summarise(total_cells = n())
save(total_cells, file = "data/total_cells.rda")

total_cells %>%
    ggplot(aes(sample,
               total_cells,
               fill = sample)) +
    geom_bar(stat = "identity") +
    geom_text(aes(label = total_cells)) +
    ggtitle("total number of cells") +
    theme(legend.position = "none") +
    ylab("cell count (10k read minimum)")
```



## Total RNA read counts

When sequencing more cells, we have overall less counts per cell. The distribution of counts here looks reasonable; we are truncating the lower end because we required there to be at least 10,000 reads to match a barcode before aligning. The count depth for the unsorted group is lower than we'd like to see.

```{r total_counts}
metrics %>%
    ggplot(aes(total_counts,
               fill = sample)) +
    facet_wrap(~sample) +
    geom_histogram() +
    ggtitle("total RNA read counts") +
    scale_x_log10() +
    scale_y_log10() +
    theme(axis.text.x = element_text(angle = 90, hjust = 1),
          legend.position = "none") +
    xlab("counts per cell")

metrics %>%
    ggplot(aes(sample,
               total_counts,
               fill = sample)) +
    geom_boxplot() +
    geom_label(data = aggregate(total_counts ~ sample, metrics, median),
              aes(label = round(total_counts))) +
    ggtitle("total RNA read counts") +
    scale_y_log10() +
    theme(legend.position = "none") +
    ylab("counts per cell")
```



## Genes detected

Here by "detected", we mean genes with a non-zero count measurement per cell. Seeing gene detection in the range of ~2,000-2,500 is normal for [inDrop][] analysis. The low gene detection rate of the unsorted group is likely problematic.

```{r genes_detected}
metrics %>%
    ggplot(aes(genes_detected,
               fill = sample)) +
    facet_wrap(~sample) +
    geom_histogram() +
    ggtitle("genes detected") +
    theme(axis.text.x = element_text(angle = 90, hjust = 1),
          legend.position = "none") +
    xlab("genes per cell")

metrics %>%
    ggplot(aes(sample,
               genes_detected,
               fill = sample)) +
    geom_boxplot() +
    geom_label(data = aggregate(genes_detected ~ sample, metrics, median),
              aes(label = round(genes_detected))) +
    ggtitle("genes detected") +
    theme(legend.position = "none") +
    ylab("genes per cell")
```



## Total vs. detected

If we graph out the total number of counts per cell vs. the genes detected per cell, we can assess whether there is a large population of low quality cells with low counts and/or gene detection. We're not seeing this in the FACS sorted group, which is good. There's more spread in the unsorted group, but not a huge shoulder of low complexity, which has been seen in some other studies.

```{r total_vs_detected}
metrics %>%
    ggplot(aes(total_counts,
               genes_detected)) +
    facet_wrap(~sample) +
    geom_point() +
    geom_smooth(method = "loess") +
    ggtitle("total counts vs. genes detected") +
    scale_x_log10() +
    scale_y_log10() +
    theme(axis.text.x = element_text(angle = 90, hjust = 1)) +
    xlab("counts per cell") +
    ylab("genes per cell")
```



## Mitochondrial counts

We evaluate overall mitochondrial gene expression as a biomarker of cellular stress during sample preparation. We're not seeing very high mitochondrial gene expression (> 20k counts) in most cells, so that's good. The difference looks large between the `sorted` and `unsorted` groups here because the count depth is relatively low in the `unsorted` group. We've seen samples in other studies with higher mitochondrial gene expression.

```{r mitochondrial_counts}
metrics %>%
    ggplot(aes(coding_counts,
               mito_counts,
               color = sample)) +
    facet_wrap(~sample) +
    ggtitle("mitochondrial gene abundance") +
    geom_point() +
    theme(axis.text.x = element_text(angle = 90, hjust = 1),
          legend.position = "none") +
    xlab("counts in mitochondrial genes") +
    xlim(0, 50000) +
    ylab("counts in coding genes")
```

```{r percent_mito}
metrics %>%
    dplyr::mutate(percent_mito = percent_mito * 100) %>%
    ggplot(aes(percent_mito,
               fill = sample)) +
    facet_wrap(~sample) +
    geom_histogram() +
    ggtitle("% mitochondrial gene abundance") +
    scale_y_sqrt() +
    theme(legend.position = "none") +
    xlab("% mitochondrial")
```



## Novelty

Another way to QC the data is to look for less novelty, that is cells that have less genes detected per count than other cells. We can see the samples where we sequenced each cell less have a higher overall novelty, that is because we have not started saturated the sequencing for any given gene for these samples. Outlier cells in these samples might be cells that we have a less complex RNA species than other cells. Sometimes we can detect contamination with low complexity cell types like red blood cells via this metric.

```{r novelty}
metrics %>%
    ggplot(aes(log_detected_per_count,
               fill = sample)) +
    facet_wrap(~sample) +
    geom_histogram() +
    ggtitle("novelty") +
    scale_y_sqrt() +
    theme(legend.position = "none") +
    xlab("log genes detected per count")

metrics %>%
    ggplot(aes(sample,
               log_detected_per_count,
               fill = sample)) +
    geom_boxplot() +
    ggtitle("novelty") +
    geom_label(data = aggregate(log_detected_per_count ~ sample, metrics, median),
              aes(label = round(log_detected_per_count, digits = 2))) +
    theme(legend.position = "none") +
    ylab("log genes detected per count")
```



# Summary

We can continue on to clustering analysis with the FACS sorted sample. The unsorted sample has low cell counts and low gene detection. As mentioned above, the FACS sorted sample has a bimodal distribution of the cellular barcodes, which is not desirable. We'll work on figuring out the reason behind this. We don't see a big shoulder of low complexity cells in the FACS group, and we recommend applying only minimal filtering based on mitochondrial content and novelty.


## Suggested filtering

We recommend proceeding with minimal filtering:

- `> 500` genes detected per cell (abnormally low)
- `< 5000` genes detected per cell (abnormally high)
- `< 20%` mitochondrial transcripts
- `> 0.8` novelty score (log genes detected per count)

```{r filtered}
filtered <- metrics %>%
    dplyr::filter(genes_detected > 500 &
                      genes_detected < 5000,
                      percent_mito < 0.20 &
                      log_detected_per_count > 0.8) %>%
    # FACS sorted sample only!
    dplyr::filter(sample == "sorted")

filtered %>%
    ggplot(aes(genes_detected,
               fill = sample)) +
    facet_wrap(~sample) +
    geom_histogram() +
    theme(axis.text.x = element_text(angle = 90, hjust = 1),
          legend.position = "none") +
    xlab("genes")

filtered %>%
    ggplot(aes(total_counts,
               genes_detected)) +
    facet_wrap(~sample) +
    geom_point() +
    geom_smooth(method = "lm") +
    scale_x_log10() +
    scale_y_log10() +
    theme(axis.text.x = element_text(angle = 90, hjust = 1)) +
    xlab("counts") +
    ylab("genes")
```

```{r}
sessionInfo()
```
