---
title: "Analysis of CpG methylation correlation with gene expression using public data"
output: pdf_document

---

```{r setupstate, echo=FALSE}
library(knitr)
opts_chunk$set(tidy=TRUE, cache=TRUE,  highlight=TRUE, figalign="center", echo=FALSE, warning=FALSE, error=FALSE, message=FALSE, tidy.opts=list(keep.blank.line=FALSE, width.cutoff=120))
options(width=200)
```

# Analysis of CpG methylation correlation with gene expression using public data

## Summary  

Analysis of public data methylation and gene expression datafor [Daniel Kim](mailto:daniel.kim@post.harvard.edu) a visting scientist with the Baccarelli group at HSPH.

Processed public data for 450K and gene expression were downloaded from GEO, subsetted to CpGs and genes of interest and examined for correaltion between gene expression at these genes and the methylation status of the CpGs in their promomters.

Contact [John Hutchinson](mailto:jhutchin@hsph.harvard.edu) for additional details.

The most recent update of this html document occurred: `r date()`

The sections below provide code to reproduce the included results and plots.

---

#  SETUP

## R libraries necessary for analysis

I used the ([GEOquery][1]) [Bioconductor][2] library to interface with the [Gene Expression Omnibus][3] database.


```{r general_libraries, echo=TRUE}
source("http://bioconductor.org/biocLite.R") # BioConductor script necessary for installing new BioC libraries with biocLite()
library(hgu133plus2.db)
library(GEOquery)
library(ggplot2) # library for plotting 
library(dplyr)
library(readr)
library(tidyr)
library(stringr)
library(reshape2)
library(R.utils)
library(knitr)
```

## Locations of directories for data, metadata and results

```{r general_directories}
baseDir <- "~/Work/projects/ab_CpGexpression"
dataDir <- file.path(baseDir, "data")
resultsDir <- file.path(baseDir, "results")
metaDir <- file.path(baseDir, "meta")
````

----

# Data Intersection

##  Illumina annotations
- load in annotations for Illunina 450K array (*HumanMethylation450_15017482_v.1.1*)
- here is an example of the data contained in these annotations
 
```{r illumnannnots}
annots <- read_delim(file.path(metaDir, "HumanMethylation450_15017482_v.1.1.nometaheader.tab"), delim="\t")
annots <- select(annots, -Forward_Sequence, -SourceSeq, -AlleleA_ProbeSeq, -AlleleB_ProbeSeq, -contains("SNP"), -contains("Channel"), -contains("Loci"), -contains("Address"), -contains("Base"), -contains("Infinium"), -Phantom, -DMR, -Enhancer, -HMM_Island, -Relation_to_UCSC_CpG_Island, -contains("Regulatory"))
kable(head(annots), caption="Subset of Illumina450K annotations")
```

## CpG locations
- provided by Daniel Kim
- all are located within promoters of respective genes
- genomic locations are for hg18 genome build

```{r loadCpGdata}
genescpgs <- read_csv(file.path(dataDir, "genes_and_CpGs.csv"))
genescpgs <- filter(genescpgs, !(is.na(Gene)))
kable(genescpgs, captions="Original format")
genescpgs <-melt(genescpgs, id.vars=c("Gene", "Chr")) %>% tbl_df() %>% filter(., !is.na(value))
genescpgs <- mutate(genescpgs, as.character(Chr))
genescpgs <- select(genescpgs, -Chr)
names(genescpgs) <- sub("as.character\\(", "", names(genescpgs)) %>% sub("\\)", "", .)
kable(genescpgs, caption="Reformatted")
```

## Common CpGs between Illumina 450K platform and experimental data
- using hg18 bulid, find Illumina 450K probeids that are within 1bp of provided CpG locations

```{r intersect}
genescpgs <- genescpgs %>% filter(., grepl("CpG", variable)) %>% arrange(Gene)

joined_annots <- inner_join(genescpgs, annots, by=c("Chr"="CHR") )

probes_of_interest <- joined_annots %>% mutate(diff=value-Coordinate_36) %>% filter(abs(diff)<2)

names(probes_of_interest) <- names(probes_of_interest) %>% sub("variable", "CpGID", .) %>% sub("value", "position", .) 
 probes_of_interest <- select(probes_of_interest, CpGID, position, Chr, IlmnID, Chromosome_36,Coordinate_36, UCSC_RefGene_Name )
 
 kable(probes_of_interest)

```

---

# Data Analysis
## By individual study

Studies were chosen by searching through the GEO datasets (http://www.ncbi.nlm.nih.gov/gds) to look for public datasets that would allow correlation of gene expression with methylation at your CpGs. To do so, I searched GEO for any datasets with:
a) data from the 450K array
b) and any type of gene expression data, whether by RNAseq or microarray
c) human derived
d) the word blood anywhere in the dataset description 

GPL13534[RGSE] AND (expression profiling by high throughput sequencing[DataSet Type] OR expression profiling by array[DataSet Type]) AND human[organism] AND blood

This yielded 39 potential datasets, which I then manually examined to see if they were relevant to your experiment (i.e. contained control blood samples).

Most of the datasets were from studies directed to very specific questions (i.e. knockout studies, stem cell development ) that weren’t relevant to the experiment but I did find a few studies that might be relevant to the reviewer’s question.

The first two are both in peripheral blood but have very small sample sizes.
1) http://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE34779
This study looked at the relationship between promoter methylation and gene expression in peripheral blood from 8 individuals. This study is probably the most directly relevant to your study.
2) http://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE67491 
This longitudinal study looked at blood samples from 12 individuals at six month intervals over the course of a year and a half. 

Two other studies didn’t look at peripheral blood, but have the advantage of a much higher numbers of samples, which would increase your chances of showing co-variation of promoter methylation and gene expression.
1) http://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE56047 
This study looked at age-related variation in methylation and expression in human monocytes and T cells.
2) http://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE53261
This study looked at methylation and expression variation in untransformed  human fibroblasts.

Given the time constraints, it’s probably best to just quickly look at the very first of these studies, (with the 8 individuals in peripheral blood). Otherwise, you might want to take the tack with the reviewer that 1) studies of DNA methylation and gene expression on the same blood samples are still rare and b) sample sizes for these studies are still very small and underpowered. 

**The IDs of the GEO series (GSEs) are:**

```{r GEOIDs}
GSE.ids <- c("GSE34779", "GSE67491","GSE56047","GSE53261")
  
print(GSE.ids)
```

**There are `r length(GSE.ids)` GEO series in total**

**Data for each GEO series was downloaded via GEOquery and analysed with plots plots and tables of correlation (spearman) values

### GSE34799
- composed of two sub-series datasets (GSE34776 -	Expression data from human lymphoblasts [Affymetrix U133Plus2.0 array], GSE34777 -DNA methylation data from human lymphoblasts [Illumina 450K array])

```{r GSE34799, GSE34779}
GSE <- "GSE34799"
exprdata <- getGEO("GSE34776", GSEMatrix = TRUE)
eset <- exprdata[[1]] %>% exprs() %>% as.data.frame()
eset$probeid <- row.names(eset)
eset <- tbl_df(eset)
eset <- eset %>% mutate(symbols=unlist(mget(probeid, hgu133plus2SYMBOL, ifnotfound = NA)))
eset <- melt(eset) %>% tbl_df()
pd.eset <- pData(exprdata[[1]]) %>% tbl_df() %>% mutate(., id=sub(" .*$", "", title)) %>% select(.,id,geo_accession )
eset <- full_join(eset, pd.eset, by=c("variable"="geo_accession"))

methdata <- getGEO("GSE34777", GSEMatrix = TRUE)
mset <- methdata[[1]] %>% exprs() %>% as.data.frame()
mset$probeid <- row.names(mset)
mset <- tbl_df(mset)
mset <- melt(mset) %>% tbl_df()
pd.mset <- pData(methdata[[1]]) %>% tbl_df() %>% mutate(., id=sub(" .*$", "", title)) %>% select(., id, geo_accession)
mset <- full_join(mset, pd.mset, by=c("variable"="geo_accession"))

# subset data to genes and cpGs of interest
eset.tlr2 <- filter(eset, symbols=="TLR2")
eset.tlr2 <- mutate(eset.tlr2, symbol_probe=paste(symbols, probeid, sep="-"))
mset.tlr2 <-  filter(mset, probeid %in% probes_of_interest$IlmnID)
# add back original CpG annotations
mset.tlr2 <- inner_join(mset.tlr2, probes_of_interest, by=c("probeid"="IlmnID"))

mset.eset.tlr2 <- inner_join(eset.tlr2, mset.tlr2, by="id") 
names(mset.eset.tlr2) <- names(mset.eset.tlr2) %>% sub(".x$", ".expr", .) %>% sub(".y$", ".meth", .)

ggplot(mset.eset.tlr2, aes(x=value.expr, y=value.meth)) + geom_point() +facet_wrap( ~ symbols+ probeid.expr + probeid.meth + CpGID+ position,ncol=1 )

correlations <- mset.eset.tlr2 %>% group_by(probeid.meth, probeid.expr) %>% summarize(.,correlation=cor(value.expr, value.meth, method="spearman")) 

inner_join(correlations, probes_of_interest, by=c("probeid.meth"="IlmnID")) %>% select(., UCSC_RefGene_Name, Chr, position, CpGID,  probeid.meth, probeid.expr,    correlation) %>% kable()

````

###GSE67491
- composed of two sub-series datasets (GSE67488- Expression data from human blood  [RNA-seq], GSE67490 - DNA methylation data from human blood [Illumina 450K])

```{r GSE67491}
GSE <- "GSE67491"

#mgetGEOSuppFiles("GSE67488", baseDir=dataDir)
#gunzip(file.path(dataDir, "GSE67488", "GSE67488_whole_12_RNAseq_snm_normalized_fpkm.txt.gz"))
eset <- read_delim(file.path(dataDir, "GSE67488","GSE67488_whole_12_RNAseq_snm_normalized_fpkm.txt"), delim="\t")
eset <- melt(eset) %>%tbl_df()
names(eset) <- c("symbols", "id.expr", "value.expr")

methdata <- getGEO("GSE67490", GSEMatrix = TRUE)
mset <- methdata[[1]] %>% exprs() %>% as.data.frame()
mset$probeid <- row.names(mset)
mset <- tbl_df(mset)
mset <- melt(mset) %>% tbl_df()
pd.mset <- pData(methdata[[1]]) %>% tbl_df() %>% mutate(., id=sub(" .*$", "", title)) %>% select(., id, geo_accession)
mset <- full_join(mset, pd.mset, by=c("variable"="geo_accession"))
names(mset) <- c("probeid.meth", "GEOid.meth", "value.meth", "id.meth")


# subset data to genes and cpGs of interest
eset.tlr2 <- filter(eset, symbols=="TLR2")
mset.tlr2 <-  filter(mset, probeid.meth %in% probes_of_interest$IlmnID)
# add back original CpG annotations
mset.tlr2 <- inner_join(mset.tlr2, probes_of_interest, by=c("probeid.meth"="IlmnID"))


mset.eset.tlr2 <- inner_join(mset.tlr2,eset.tlr2,  by=c("id.meth"="id.expr"))


ggplot(mset.eset.tlr2, aes(x=value.expr, y=value.meth)) + geom_point() +facet_wrap( ~ symbols+ probeid.meth + CpGID+ position,ncol=1 )

correlations <- mset.eset.tlr2 %>% group_by(probeid.meth, symbols) %>% summarize(.,correlation=cor(value.expr, value.meth,method="spearman")) 

inner_join(correlations, probes_of_interest, by=c("probeid.meth"="IlmnID")) %>% select(., UCSC_RefGene_Name, Chr, position, CpGID,  probeid.meth, symbols,    correlation) %>% kable()

```

[1]: http://www.bioconductor.org/packages/2.11/bioc/html/GEOquery.html (GEOquery_BioC_library)
[2]: http://www.bioconductor.org (BioC)
[3]: http://www.ncbi.nlm.nih.gov/gds/ (GEO)


