---
title: "Cell deconvolutions - Normal lungs and lungs with distal tumors"
author: "`r getOption('author')`"
date: "`r Sys.Date()`"
bibliography: bibliography.bib
params:
    bcbFile: "data/bcb_sub_lung.rda"
    outputDir: "."
---

```{r setup, message=FALSE}
library(bcbioRNASeq)
library(DeconRNASeq)
library(rio)


# Shared R Markdown settings
prepareRNASeqTemplate()
if (file.exists("setup.R")) {
    source("setup.R")
}

# Directory paths
dataDir <- file.path(params$outputDir, "data")

# Load bcbioRNASeq object
bcbName <- load(params$bcbFile)
bcb <- get(bcbName, inherits = FALSE)


kable(as.data.frame(colData(bcb)))
```

```{r header, child="_header.Rmd", eval=file.exists("_header.Rmd")}
```

The idea here is to compare the ratio of neutrophil gene expression cell signatures (up or down in KEP cells from Coffelt paper) in our normal lung and lungs with distal tumors to see if there are any shifts.

We use the TPM matrix from the bcbio run for this as it is library and gene size normalized. 

```{r counts, include=FALSE}
tpm <- tpm(bcb) %>% as.data.frame()
rowData <- rowData(bcb)

annotated <- cbind(as.data.frame(tpm), as.data.frame(rowData))
```

# Coffelt signatures

Import top 50 differentially expressed genes from Coffelt et al., 2015 (Nature). (from extended Table 1).

Subset to genes with pvalues <0.1

```{r signatures}
signatures <- import("~/Orchestra/PIs/sandra_mcallister/lung_metastasis/meta/coffelt_2015_extended_table1_top50DEgenes.csv")

kable(signatures)

kepup_genes <- subset(signatures, pvalue<0.1 & direction=="positive")$ensemblid
kepdown_genes <- subset(signatures, pvalue<0.1 & direction=="negative")$ensemblid


kable(kepup_genes)
```

Get the ratio of the geometric means for the signature genes for all samples in the dataset.

```{r geom_means}
gm_mean = function(x, na.rm=TRUE){
  exp(sum(log(x[x > 0]), na.rm=na.rm) / length(x))
}

tpm_kepup <- tpm[kepup_genes,] 
tpm_kepup <- tpm_kepup[!apply(tpm_kepup, 1, function(x) all(is.na(x))),]

tpm_kepdown <- tpm[kepdown_genes,] 
tpm_kepdown <- tpm_kepdown[!apply(tpm_kepdown, 1, function(x) all(is.na(x))),]

kepdownmeans <- tpm_kepdown %>% apply(., 2, function(x) gm_mean(x))
kepupmeans <- tpm_kepup %>% apply(., 2, function(x) gm_mean(x))
ratios <- kepupmeans/kepdownmeans %>% as.data.frame() 
ratios <-rownames_to_column(ratios, var="sampleID")
names(ratios) <- c("sampleID", "ratio")
```

Add group ids to data for plotting.
- remove anything with a RIN below 3

```{r annotateratios1}
ratios_annot  <- inner_join(ratios, as.data.frame(colData(bcb)))

# remove anything with a RIN below 2.5

ratios_annot$rin <-   as.numeric(as.character(ratios_annot$rin))
ratios_annot <- subset(ratios_annot, rin>3)

p <- ggplot(ratios_annot, aes(x=sampleclass, y=ratio, color=sampleclass))+geom_boxplot()
p +  geom_jitter(shape=16, size=3 , position=position_jitter(0.00))+ggtitle("Coffelt signature")

t.test(subset(ratios_annot, ratios_annot$sampleclass=="lung_distaltumor" & ratios_annot$rin>3)$ratio,subset(ratios_annot, ratios_annot$sampleclass=="lung_notumor" & ratios_annot$rin>3)$ratio)

```


