---
output:
  html_document:
    code_folding: hide
    theme: yeti
    toc: yes
    toc_float: true
    toc_depth: 4
    number_sections: true
    fig_height: 6
---


```{r setup1, echo=FALSE}
# Setup report details
clientname="Jaclyn Sceneay"
clientemail="JSCENEAY@partners.org"
lablocation="Sandra McAllister" 
analystname="John Hutchinson"
analystemail="jhutchin@hsph.harvard.edu"
```

> RNA-Seq analysis for `r clientname` (`r clientemail`) at `r lablocation`. Contact `r analystname` (`r analystemail`) for additional details. 
> This document was generated on `r date()`
 
```{r setup2, echo=FALSE}
knitr::opts_chunk$set(tidy=TRUE, highlight=TRUE, dev="png",
               cache=TRUE, highlight=TRUE, autodep=TRUE, warning=FALSE, error=FALSE,
               message=FALSE, prompt=TRUE, comment='', fig.cap='')
```

# Overview

> The project has 60 samples from mouse tumors. Samples were from young and old mice and were treated with one of three different treatments (two immunotherapies plus isotype control). Multiple controls are also included.

Here I do some basic QC to see if the data looks good and if whether any samples look like they have issues.

```{r check-library-functions, echo=FALSE}
check_install = function(packages) {
   not_installed = setdiff(packages, rownames(installed.packages()))
   if(length(not_installed) > 0) {
      write(paste("The libraries", not_installed, "are not installed, aborting.",
                   sep=" "), stdout())
      stop()
   }
}
```

```{r load-libraries, echo=FALSE}
packages = c("ggplot2", "reshape", "gplots", "edgeR", "CHBUtils", "pheatmap",
              "DESeq2", "tximport", "DT", "DEGreport", "dplyr", "rio", "janitor")
check_install(packages)
installed = lapply(packages, library, character.only = TRUE)
```

```{r qc-setup}
project_summary = "/Volumes/orch/group_dir/PIs/sandra_mcallister/jaclyn-tumor-rnaseq/bcbio/final/2017-03-02_bcbio/project-summary.csv"
counts_file = "/Volumes/orch/group_dir/PIs/sandra_mcallister/jaclyn-tumor-rnaseq/bcbio/final/2017-03-02_bcbio/combined.counts"
tx2genes_file = "/Volumes/orch/group_dir/PIs/sandra_mcallister/jaclyn-tumor-rnaseq/bcbio/final/2017-03-02_bcbio/tx2gene.csv"
cbPalette <- c("#999999", "#E69F00", "#56B4E9", "#009E73", "#F0E442","#0072B2", "#D55E00", "#CC79A7")
summarydata = import(project_summary)
summarydata = summarydata[,colSums(is.na(summarydata)) < nrow(summarydata)]
# handle newer bcbio-nextgen runs that use description as the key
if("description" %in% colnames(summarydata)) {
  rownames(summarydata) = summarydata$description
  summarydata$name = rownames(summarydata)
  summarydata$description = NULL
} else {
  rownames(summarydata) = summarydata$name
  # summarydata$name = NULL
}
summarydata = summarydata[order(rownames(summarydata)),]
if (file.exists(tx2genes_file)) {
  sample_dirs = file.path("/Volumes/orch/projects/jaclyn-tumor-rnaseq/data/",  rownames(summarydata))
  salmon_files = file.path(sample_dirs, "salmon", "quant.sf")
  sailfish_files = file.path(sample_dirs, "sailfish", "quant.sf")
  new_sailfish = file.path(sample_dirs, "sailfish", "quant", "quant.sf")
  new_salmon = file.path(sample_dirs, "salmon", "quant", "quant.sf")
  if (file.exists(salmon_files[1])) {
    sf_files = salmon_files
  } else if (file.exists(sailfish_files[1])) {
    sf_files = sailfish_files
  }
  else if (file.exists(new_sailfish[1])) {
    sf_files = new_sailfish
  }
  else if (file.exists(new_salmon[1])) {
    sf_files = new_salmon
  }
  names(sf_files) = rownames(summarydata)
  tx2gene = read.table(tx2genes_file, sep=",", row.names=NULL, header=FALSE)
  txi.salmon = tximport(sf_files, type="salmon", tx2gene=tx2gene,
                        reader=readr::read_tsv, countsFromAbundance="lengthScaledTPM")
  counts = round(data.frame(txi.salmon$counts, check.names=FALSE))
} else {
  counts = read.table(counts_file, header=TRUE, row.names="id", check.names=FALSE)
}

counts = counts[, order(colnames(counts)), drop=FALSE]
colnames(counts) = gsub(".counts", "", colnames(counts))
save(file="RDAta.counts", counts)

# this is a list of all non user-supplied metadata columns that could appear
known_columns = c("name", "X.GC", "Exonic.Rate", "Sequences.flagged.as.poor.quality",
    "rRNA_rate", "Fragment.Length.Mean", "Intronic.Rate", "Intergenic.Rate",
    "Mapping.Rate", "Quality.format", "Duplication.Rate.of.Mapped", "Mapped",
    "rRNA", "Sequence.length", "Transcripts.Detected", "Mean.Per.Base.Cov.",
    "Genes.Detected", "Unique.Starts.Per.Read", "unique_starts_per_read",
    "complexity", "X5.3.bias", "Duplicates.pct", "Duplicates", "Mapped.reads",
    "Average.insert.size", "Mapped.reads.pct", "Total.reads",
    "avg_coverage_per_region", "Mapped.Reads", "Quality format", "Mapped_reads","Intergenic Rate", "Duplication Rate of Mapped", "Mapped_paired_reads", "Sequence length", "Exonic Rate", "Total_reads", "%GC", "Sequences flagged as poor quality", "Average_insert_size", "Intronic Rate", "5'-3' bias", "Fragment_Length_Mean", "Mapped_reads_pct", "rrna_rate","duplication_rate_of_mapped","mapped_reads","mapped_paired_reads","quality_format","intergenic_rate","sequence_length","exonic_rate","total_reads","percentgc","rrna","duplicates","sequences_flagged_as_poor_quality","average_insert_size","intronic_rate","x5_3_bias","fragment_length_mean","mapped_reads_pct")
summarydata[,"Fragment_Length_Mean"] = summarydata$Average_insert_size

summarydata[,"Mapped_reads_pct"] <- (summarydata$Mapped_reads/summarydata$Total_reads)*100

summarydata <- clean_names(summarydata)
metadata = summarydata[, !colnames(summarydata) %in% known_columns, drop=FALSE]
metadata = metadata[, colSums(is.na(metadata)) < nrow(metadata), drop=FALSE]
```

```{r sanitize-datatable}
sanitize_datatable = function(df, ...) {
 # remove dashes which cause wrapping
 DT::datatable(df, ..., rownames=gsub("-", "_", rownames(df)),
                   colnames=gsub("-", "_", colnames(df)))
}
```

```{r set-random-seed}
# set seed for reproducibility
set.seed(1454944673)
```

## Sample metadata
- various experimental and technical factors

```{r metadata-table, echo=FALSE, cache=FALSE}
sanitize_datatable(metadata, style='bootstrap')
```

# Quality control metrics

## Sample data metrics
- some metrics that tell us how well the sequencing worked

```{r summary-table, echo=FALSE, cache=FALSE}
sanitize_datatable(summarydata, style='bootstrap')
```

```{r heatmap-function}
get_heatmap_fn = function(summarydata) {
    # return the pheatmap function with or without metadata
    if(ncol(metadata) == 0) {
       return(pheatmap)
    }
    else {
    # rownames(metadata) = summarydata$name
    heatmap_fn = function(data, ...) {
        pheatmap(data, annotation=metadata, clustering_method = "ward.D2",
        clustering_distance_cols = "correlation", ...)
    }
    return(heatmap_fn)
}}
heatmap_fn = get_heatmap_fn(summarydata)
```


```{r qualimap-run}
qualimap_run = "mapped" %in% colnames(summarydata)
do_quality = "mapped_reads" %in% colnames(summarydata)
```

## Mapped reads

Just the number of reads (with pairing factored in, i.e. pairs count for one read) that aligned uniquely to the genome.

```{r mapped-plot-noqualimap, eval=!qualimap_run & do_quality}
ggplot(summarydata, aes(x=name, y=mapped_reads)) +
    theme_bw(base_size=10) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=90)) +
    geom_bar(stat="identity") +
    ylab("mapped reads") + xlab("")
```

The number of mapped reads is lower than we'd like to see, but biological replicates may make up for some of the missing data.

## Genomic mapping rate

What percentage of the total reads sequenced actually align.

```{r mapping-rate-plot-noqualimap, eval=!qualimap_run & do_quality}
ggplot(summarydata, aes(x=name, y=mapped_reads_pct)) +
    geom_bar(stat="identity") +
    ylab("mapping rate") + xlab("") +
    theme_bw(base_size=10) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=90))
```

Overall pretty good, will keep an eye on the 5/6 samples with lower rates.

## Number of genes detected

How many genes can we "detect". i.e. for how many genes do we have a minimum number of reads present?

```{r genes-detected-plot}
dd = data.frame(name=colnames(counts), genes_detected = colSums(counts > 0))
ggplot(dd, aes(x=name, y=genes_detected)) +
    geom_bar(stat="identity") +
    theme_bw(base_size=10) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=90)) +
    ylab("genes detected") + xlab("")
```

Detection rates are good, perfectly in line with what you'd expect for a mouse sequencing run.

## Gene detection saturation
```{r saturation-plot, eval=do_quality}
col_mapped = ifelse(qualimap_run, "mapped", "mapped_reads")
dd = data.frame(mapped=summarydata[,col_mapped], genes_detected = colSums(counts > 0), treatment=summarydata$treatment_short,age=summarydata$age)
ggplot(dd, aes(x=mapped, y=genes_detected, color=treatment)) +
    geom_point(size=5) +
    theme_bw(base_size=10) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=90)) +
    ylab("genes detected") + xlab("reads mapped")
```

The various control samples do appear to have typically different numbers of genes detected and mapped.
The red dot is the one sample I have a fastq data for but no metadata.

## Exonic mapping rate
```{r exonic-mapping-plot, eval="exonic_rate" %in% colnames(summarydata)}
ggplot(summarydata, aes(x=name, y=exonic_rate)) +
    geom_bar(stat="identity") +
    theme_bw(base_size=10) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=90)) +
    ylab("exonic mapping rate") + xlab("")
```

Exonic mapping rate looks really good.

## rRNA mapping rate
```{r check-rRNA-eval}
eval_rRNA = "rrna_rate" %in% colnames(summarydata) & !sum(is.na(summarydata$rrna_rate)) == nrow(summarydata)
```
```{r rRNA-rate-plot, eval=eval_rRNA & do_quality}
ggplot(summarydata, aes(x=name, y=rrna_rate)) +
    geom_bar(stat="identity") +
    theme_bw(base_size=10) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=90)) +
    ylab("rRNA rate") + xlab("")
```

All of these rRNA rates are really good (i.e. far less than 0.1%), we are generally OK with anything below 5%.

## Estimated fragment length of paired-end reads
```{r fragment-length-plot, eval="Fragment_Length_Mean" %in% colnames(summarydata) & do_quality}
ggplot(summarydata, aes(x=name, y=fragment_length_mean)) +
    geom_bar(stat="identity") +
    theme_bw(base_size=10) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=90)) +
    ylab("fragment length") + xlab("")
```

## 5'->3' bias
```{r 53bias}
ggplot(summarydata, aes(x=name, y=x5_3_bias)) +
    geom_bar(stat="identity") +
    theme_bw(base_size=10) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=90)) +
    ylab("5'->3' bias") + xlab("")
```

## Boxplot of log10 counts per gene
```{r boxplot-raw}
melted = melt(counts)
colnames(melted) = c("sample", "count")
melted$sample = factor(melted$sample)
melted = melted[order(melted$sample),]
melted$count = log(melted$count)
ggplot(melted, aes(x=sample, y=count)) + geom_boxplot() +
    theme_bw(base_size=10) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=90)) + xlab("")
```

## Boxplot of log10 TMM-normalized counts per gene
Trimmed mean of M-values (TMM) normalization is described
[here](http://genomebiology.biomedcentral.com/articles/10.1186/gb-2010-11-3-r25)

Robinson, M. D., & Oshlack, A. (2010). A scaling normalization method for differential expression analysis of RNA-seq data. Genome Biology, 11(3). doi:10.1186/gb-2010-11-3-r25

```{r boxplot-normalized}
y = DGEList(counts=counts)
y = calcNormFactors(y)
normalized_counts = cpm(y, normalized.lib.sizes=TRUE)
melted = melt(normalized_counts)
colnames(melted) = c("gene", "sample", "count")
melted$sample = factor(melted$sample)
melted = melted[order(melted$sample),]
melted$count = log(melted$count)
ggplot(melted, aes(x=sample, y=count)) + geom_boxplot() +
    theme_bw(base_size=10) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=90)) + xlab("")
```

## Density of log10 TMM-normalized counts
```{r density-normalized}
ggplot(melted, aes(x=count, group=sample)) +
    geom_density() +
    theme_bw(base_size=10) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=90)) + xlab("")
```

The count distributions all look consistent, which is good.

## Correlation heatmap of TMM-normalized counts {.tabset}

### Correlation (Pearson)
```{r pearson-heatmap-normalized1, eval=ncol(counts) > 1, fig.height=12, fig.width=14}
pheatmap(cor(normalized_counts, method="pearson"), annotation=metadata[,c("age", "treatment", "cage_num", "rna_sample_num", "mfp_side", "response")])
```

### Correlation (Spearman)
```{r spearman-heatmap-normalized1, eval=ncol(counts) > 1, fig.height=12, fig.width=14}
pheatmap(cor(normalized_counts, method="spearman"), annotation=metadata[,c("age", "treatment", "cage_num", "rna_sample_num", "mfp_side", "response")])
```

## Correlation heatmap of TMM-normalized counts without controls{.tabset}

The controls are definitely different from the rest of the samples. 

I tried seeing what the correlation plots look like without those samples.

### Correlation (Pearson)
```{r pearson-heatmap-normalized2, eval=ncol(counts) > 1, fig.height=12, fig.width=14}
normalized_counts.sub <- normalized_counts[,which(metadata$general_class=="exp")]
metadata.sub <- metadata[which(metadata$general_class=="exp"),]
pheatmap(cor(normalized_counts.sub, method="pearson"), annotation=metadata.sub[,c("age", "treatment", "cage_num", "rna_sample_num", "mfp_side", "response")])
```

### Correlation (Spearman)
```{r spearman-heatmap-normalized2, eval=ncol(counts) > 1, fig.height=12, fig.width=14}
pheatmap(cor(normalized_counts.sub, method="spearman"), annotation=metadata.sub[,c("age", "treatment", "cage_num", "rna_sample_num", "mfp_side", "response")])
```

Still not seeing any great deal of clustering, but maybe the PCA will show something.

## PCA plots

### All samples {.tabset}

```{r pca, eval=ncol(counts) > 1}
dds = DESeqDataSetFromMatrix(countData=counts, colData=summarydata, design=~name)
vst = varianceStabilizingTransformation(dds)
pca_loadings = function(object, ntop=500) {
  rv <- matrixStats::rowVars(assay(object))
  select <- order(rv, decreasing = TRUE)[seq_len(min(ntop,
      length(rv)))]
  pca <- prcomp(t(assay(object)[select, ]))
  percentVar <- pca$sdev^2/sum(pca$sdev^2)
  names(percentVar) = colnames(pca$x)
  pca$percentVar = percentVar
  return(pca)}
pc = pca_loadings(vst)
comps = data.frame(pc$x)
comps$name = rownames(comps)
library(dplyr)
comps = comps %>% left_join(summarydata, by=c("name"="name"))
```

```{r pca-plotter}
pca_plot = function(comps, nc1, nc2, colorby) {
   c1str = paste0("PC", nc1)
   c2str = paste0("PC", nc2)
  ggplot(comps, aes_string(c1str, c2str, color=colorby)) +
    geom_point(size=2) + theme_bw() +
    xlab(paste0(c1str, ": ", round(pc$percentVar[nc1] * 100), "% variance")) +
    ylab(paste0(c2str, ": ", round(pc$percentVar[nc2] * 100), "% variance"))
  }
```

#### PC1 vs. PC2
```{r PC1-PC2}
pca_plot(comps, 1, 2, colorby="age")
pca_plot(comps, 1,2, colorby="treatment_short")
```

There is good clustering by age, but not by treatment.

#### PC3 vs. PC4
```{r PC3-PC4}
pca_plot(comps, 3, 4, colorby="age")
pca_plot(comps, 3,4, colorby="treatment_short")
```

#### PC5 vs. PC6
```{r PC5-PC6}
pca_plot(comps, 5, 6, colorby="age")
pca_plot(comps, 5,6, colorby="treatment_short")
```

#### Variance explained by component
```{r component-loadings}
ggplot(data.frame(component=reorder(names(pc$percentVar), -pc$percentVar),
                  percent_var=pc$percentVar),
       aes(component, percent_var)) + geom_bar(stat='identity') +
  ylab("percent of total variation") + xlab("") + theme_bw()
```

### Cell subsets

These are still not illuminating the issue well, will drop controls and split up young and old samples for now to see if that makes things clearer.

#### Young {.tabset}

```{r pca-young}
metadata.tb <- metadata %>% tbl_df
metadata.tb.young = metadata.tb %>% filter(age=="Young" & general_class=="exp")
sampleids.young <- metadata.tb.young$sampleid
summarydata.young <- summarydata[sampleids.young,]
vst.young <- vst[,sampleids.young]  

pc.young = pca_loadings(vst.young)
comps.young = data.frame(pc.young$x)
comps.young$name = rownames(comps.young)

comps.young = comps.young %>% left_join(summarydata.young, by=c("name"="name"))
```

##### PC1 vs. PC2
```{r PC1-PC2.yng}
nc1 <- 1
nc2 <- 2
c1str = paste0("PC", nc1)
c2str = paste0("PC", nc2)
ggplot(comps.young, aes(PC1, PC2, color=treatment_short, shape=response, size=mapped_reads)) +
  geom_point() + theme_bw() +
  xlab(paste0(c1str, ": ", round(pc$percentVar[nc1] * 100), "% variance")) +
  ylab(paste0(c2str, ": ", round(pc$percentVar[nc2] * 100), "% variance"))
```

There is some clustering for the CTLA4 samples by response, but not so much for the PDL1 samples.

It's a little concerning that most of  the PR CTLA4 samples that segregate have higher amounts of reads.

##### PC3 vs. PC4
```{r PC3-PC4.yng}
nc1 <- 3
nc2 <- 4
c1str =paste0("PC", nc1)
c2str = paste0("PC", nc2)
ggplot(comps.young, aes(PC3, PC4, color=treatment_short, shape=response, size=mapped_reads)) +
  geom_point() + theme_bw() +
  xlab(paste0(c1str, ": ", round(pc$percentVar[nc1] * 100), "% variance")) +
  ylab(paste0(c2str, ": ", round(pc$percentVar[nc2] * 100), "% variance"))

```

##### PC5 vs. PC6
```{r PC5-PC6.yng}
nc1 <- 5
nc2 <- 6
c1str = paste0("PC", nc1)
c2str = paste0("PC", nc2)
ggplot(comps.young, aes(PC5, PC6, color=treatment_short, shape=response, size=mapped_reads)) +
  geom_point() + theme_bw() +
  xlab(paste0(c1str, ": ", round(pc$percentVar[nc1] * 100), "% variance")) +
  ylab(paste0(c2str, ": ", round(pc$percentVar[nc2] * 100), "% variance"))
```

##### Variance explained by component
```{r component-loadings.yng}
ggplot(data.frame(component=reorder(names(pc.young$percentVar), -pc.young$percentVar),
                  percent_var=pc.young$percentVar),
       aes(component, percent_var)) + geom_bar(stat='identity') +
  ylab("percent of total variation") + xlab("") + theme_bw()
```


#### Old {.tabset}

```{r pca-old}
metadata.tb <- metadata %>% tbl_df
metadata.tb.old = metadata.tb %>% filter(age=="Old" & general_class=="exp")
sampleids.old <- metadata.tb.old$sampleid
summarydata.old <- summarydata[sampleids.old,]
vst.old <- vst[,sampleids.old]  

pc.old = pca_loadings(vst.old)
comps.old = data.frame(pc.old$x)
comps.old$name = rownames(comps.old)

comps.old = comps.old %>% left_join(summarydata.old, by=c("name"="name"))
```

##### PC1 vs. PC2
```{r PC1-PC2.old}
nc1 <- 1
nc2 <- 2
c1str = paste0("PC", nc1)
c2str = paste0("PC", nc2)
ggplot(comps.old, aes(PC1, PC2, color=treatment_short, shape=response, size=mapped_reads)) +
  geom_point() + theme_bw() +
  xlab(paste0(c1str, ": ", round(pc.old$percentVar[nc1] * 100), "% variance")) +
  ylab(paste0(c2str, ": ", round(pc.old$percentVar[nc2] * 100), "% variance"))
```

It looks like PC2 is capturing the difference between iso and treatment, not sure what PC1 is capturing.

##### PC3 vs. PC4
```{r PC3-PC4.old}
nc1 <- 3
nc2 <- 4
c1str =paste0("PC", nc1)
c2str = paste0("PC", nc2)
ggplot(comps.old, aes(PC3, PC4, color=treatment_short, shape=response, size=mapped_reads)) +
  geom_point() + theme_bw() +
  xlab(paste0(c1str, ": ", round(pc$percentVar[nc1] * 100), "% variance")) +
  ylab(paste0(c2str, ": ", round(pc$percentVar[nc2] * 100), "% variance"))

```

##### PC5 vs. PC6
```{r PC5-PC6.old}
nc1 <- 5
nc2 <- 6
c1str = paste0("PC", nc1)
c2str = paste0("PC", nc2)
ggplot(comps.old, aes(PC5, PC6, color=treatment_short, shape=response, size=mapped_reads)) +
  geom_point() + theme_bw() +
  xlab(paste0(c1str, ": ", round(pc$percentVar[nc1] * 100), "% variance")) +
  ylab(paste0(c2str, ": ", round(pc$percentVar[nc2] * 100), "% variance"))
```

##### Variance explained by component
```{r component-loadings.old}
ggplot(data.frame(component=reorder(names(pc.old$percentVar), -pc.old$percentVar),
                  percent_var=pc.old$percentVar),
       aes(component, percent_var)) + geom_bar(stat='identity') +
  ylab("percent of total variation") + xlab("") + theme_bw()
```

