---
output:
  html_document:
    toc: true
title: "450K probes upstream of lncRNA genes"
theme: spacelab
highlight: textmate
---

# Setup

## Load libraries
```{r libraries}
library(biomaRt)
library(knitr)
library(stringr)
library(dplyr)
```

## Functions
```{r functions}
bedTools.2in<-function(functionstring="bedtools intersect",bed1,bed2,opt.string="") {
  #create temp files
  a.file=tempfile()
  b.file=tempfile()
  out   =tempfile()
  options(scipen =99) # not to use scientific notation when writing out
 
  #write bed formatted dataframes to tempfile
  write.table(bed1,file=a.file,quote=F,sep="\t",col.names=F,row.names=F)
  write.table(bed2,file=b.file,quote=F,sep="\t",col.names=F,row.names=F)
 
  # create the command string and call the command using system()
  command=paste(functionstring,"-a",a.file,"-b",b.file,opt.string,">",out,sep=" ")
  cat(command,"\n")
  try(system(command))
 
  res=read.table(out,header=F)
  unlink(a.file);unlink(b.file);unlink(out)
  return(res)
}
bedTools.flank<-function(functionstring="bedtools flank", genomefile, bed1, opt.string="") {
  #create temp files
  a.file <- tempfile()
  out  <- tempfile()
  options(scipen=99) # not to use scientific notation when writing out
  
  #write bed formatted dataframes to tempfile
  write.table(bed1,file=a.file,quote=F,sep="\t",col.names=F,row.names=F)
  
  # create the command string and call the command using system()
  command <- paste(functionstring,"-i",a.file,"-g", genomefile, opt.string,">",out,sep=" ")
  cat(command,"\n")
  try(system(command))
  
  res <- read.table(out,header=F)
  unlink(a.file);unlink(out)
  return(res)
  }

```

## Set directories  

```{r directories}
baseDir <- "~/Work/projects/ab_lncRNAprobes/"
dataDir <- file.path(baseDir, "data")
resultsDir <- file.path(baseDir, "results")
metaDir <- file.path(baseDir, "meta")
```

## Load data
- using bed files of ncRNA positions from 
  a)  Lincipedia version 3.1 high confidence set (http://www.lncipedia.org/download)
  b) NONCODE version 4 human (http://www.noncode.org/download.php) md5=39d6c7de391dc3f83cc3cc59da573ab6

```{r dataload}
annots.450k <- read.csv(file.path(dataDir, "14_April_2014_Updated_humanmethylation450_15017482_v1-2.csv"), skip=7)
lncrna.pos <- read.delim(file.path(dataDir, "lncipedia_3_1_hc.bed"), header=F, sep="\t")
names(lncrna.pos) <- c("chr", "start", "end", "id", "score", "strand")
ncrna.pos <- read.delim(file.path(dataDir, "NONCODEv4_hg19.lncAndGene.bed","NONCODEv4u1_human_lncRNA_Gene.bed" ), header=F, sep="\t")
names(ncrna.pos) <- c("chr", "start", "end", "id", "score", "strand")
```

---

# Analysis

## Subset ncRNA genes to well annotated regions
- dicard all lncrnachondrial, patches and polymorphic regions

```{r subset}
chrs <- paste("chr", c(seq(1,22), "X", "Y"), sep="")
lncrna.pos <- lncrna.pos[lncrna.pos$chr %in% chrs,]
ncrna.pos <- ncrna.pos[ncrna.pos$chr %in% chrs,]
```

## ncRNA Genes Locations

```{r printlncrnagenes, results='asis'}
#kable(lncrna.pos)
write.table( lncrna.pos, file=file.path(resultsDir, "lincipedia.ncrna.gene.locations.xls"), sep="\t", quote=F, row.names=F, col.names=T)
write.table(ncrna.pos, file=file.path(resultsDir, "NONCODE.ncrna.gene.locations.xls"), sep="\t", quote=F, row.names=F, col.names=T)

```

### [Lincipedia gene locations](../results/lincipedia.ncrna.gene.locations.xls)
### [NONCODE gene locations](../results/NONCODE.ncrna.gene.locations.xls)


## Get coordinates of regions upstream of lncrnameric function genes  
3kb upstream of the transcript start

```{r getpromoters}
lncrna.pos.flanked <- bedTools.flank(bed1 = lncrna.pos, opt.string = "-l 3000 -s -r 0", genomefile ="~/Work/projects/ab_lncRNAprobes/meta/hg19.genome")
ncrna.pos.flanked <- bedTools.flank(bed1 = ncrna.pos, opt.string = "-l 3000 -s -r 0", genomefile ="~/Work/projects/ab_lncRNAprobes/meta/hg19.genome")

```

## Find 450K probes upstream of ncRNA genes

## Extract coordinates (Human Genome build GRCh37) of 450K probes from annotation file and convert to BED format

```{r 450kextract}
bed.450K <- annots.450k[,c("CHR", "MAPINFO", "Name")]
# adjust to zero-based coordinate format
bed.450K$start <- bed.450K$MAPINFO-1
names(bed.450K) <- c("chr", "end", "name", "start")
bed.450K$chr <- paste("chr", bed.450K$chr, sep="")
bed.450K <- bed.450K[,c("chr", "start", "end", "name")]
```

## Overlap 450K probe coordinates with regions upstream of ncRNA genes

```{r overlap}
probes450K.in.flanked.lncrnas <- bedTools.2in(bed1=bed.450K,bed2=lncrna.pos.flanked,opt.string="-wo")
probes450K.in.flanked.lncrnas <- probes450K.in.flanked.lncrnas[,c(1,2,3,4,8)]
names(probes450K.in.flanked.lncrnas) <- c("chr", "start", "end", "probeID", "upstream_of_ncrna")
# sort and remove duplicates
probes450K.in.flanked.lncrnas <- probes450K.in.flanked.lncrnas[with(probes450K.in.flanked.lncrnas, order(chr, start, probeID, upstream_of_ncrna)), ]
probes450K.in.flanked.lncrnas <- probes450K.in.flanked.lncrnas[!duplicated(probes450K.in.flanked.lncrnas),]
probes450K.in.flanked.lncrnas <- probes450K.in.flanked.lncrnas %>% tbl_df() %>% inner_join(., annots.450k, by=c("probeID"="IlmnID"))


probes450K.in.flanked.ncrnas <- bedTools.2in(bed1=bed.450K,bed2=ncrna.pos.flanked,opt.string="-wo")
probes450K.in.flanked.ncrnas <- probes450K.in.flanked.ncrnas[,c(1,2,3,4,8)]
names(probes450K.in.flanked.ncrnas) <- c("chr", "start", "end", "probeID", "upstream_of_ncrna")
# sort and remove duplicates
probes450K.in.flanked.ncrnas <- probes450K.in.flanked.ncrnas[with(probes450K.in.flanked.ncrnas, order(chr, start, probeID, upstream_of_ncrna)), ]
probes450K.in.flanked.ncrnas <- probes450K.in.flanked.ncrnas[!duplicated(probes450K.in.flanked.ncrnas),]
probes450K.in.flanked.ncrnas <- probes450K.in.flanked.ncrnas %>% tbl_df() %>% inner_join(., annots.450k, by=c("probeID"="IlmnID"))

```

---

# Results

## Overlapn

- how many common 450K probes are there between the two sets of genes

```{r venn}

LINCIPEDIA=length(unique(probes450K.in.flanked.lncrnas$probeID))
NONCODE=length(unique(probes450K.in.flanked.ncrnas$probeID))

BOTH=length(intersect(unique(probes450K.in.flanked.lncrnas$probeID), unique(probes450K.in.flanked.ncrnas$probeID)))

library(venneuler)

plot(venneuler(c("LINCIPEDIA"=LINCIPEDIA, "NONCODE"=NONCODE, "LINCIPEDIA&NONCODE"=BOTH)))
```

## Output to file

```{r output}
#kable(probes450K.in.flanked.lncrnas, row.names=F)
write.table(probes450K.in.flanked.lncrnas, file=file.path(resultsDir, "probes450K.in.flanked.lncrnas.xls"), quote=F, sep="\t", row.names=F, col.names=T)
write.table(probes450K.in.flanked.ncrnas, file=file.path(resultsDir, "probes450K.in.flanked.ncrnas.xls"), quote=F, sep="\t", row.names=F, col.names=T)
```


### [Excel file with results for Lincipedia ncRNAs](../results/probes450K.in.flanked.lncrnas.xls)
### [Excel file with results for NONCODe ncRNAs](../results/probes450K.in.flanked.ncrnas.xls)





