---
title: "Quality Control - combined runs"
author: "`r getOption('author')`"
date: "`r Sys.Date()`"
bibliography: bibliography.bib
params:
    bcb_file: "data/2018-03-13/bcb.rda"
    output_dir: "."
---

```{r setup, message=FALSE}
# Last modified 2018-02-14
bcbioRNASeq::prepareRNASeqTemplate()
source("_setup.R")

# Directory paths
data_dir <- file.path(params$output_dir, "data", Sys.Date())
results_dir <- file.path(params$output_dir, "results", "counts")
invisible(mapply(
    FUN = dir.create,
    list(data_dir, results_dir),
    MoreArgs = list(recursive = TRUE, showWarnings = FALSE)
))

# Load bcbioRNASeq object
bcb_name <- load(params$bcb_file)
bcb <- get(bcb_name, inherits = FALSE)
stopifnot(is(bcb, "bcbioRNASeq"))
validObject(bcb)
interestingGroups(bcb) <- c("treatment", "cellPopulation")
```

```{r header, child="_header.Rmd"}
```



```{r sample_metadata}
sampleMetadata(bcb)
```

[bcbio][] run data was imported from **`r metadata(bcb)$uploadDir`**.



# Count matrices

```{r counts}
raw_counts <- counts(bcb, normalized = FALSE)
normalized_counts <- counts(bcb, normalized = TRUE)
tpm <- counts(bcb, normalized = "tpm")
saveData(raw_counts, normalized_counts, tpm, dir = data_dir)
writeCounts(raw_counts, normalized_counts, tpm, dir = results_dir)
```

The results are saved as gzip-compressed comma separated values (CSV). Gzip compression is natively supported on [macOS][] and Linux-based operating systems. If you're running Windows, we recommend installing [7-Zip][]. CSV files can be opened in [Excel][] or [RStudio][].

- [`normalized_counts.csv.gz`](`r file.path(results_dir, "normalized_counts.csv.gz")`): Use to evaluate individual genes and/or generate plots. These counts are normalized for the variation in sequencing depth across samples.
- [`tpm.csv.gz`](`r file.path(results_dir, "tpm.csv.gz")`): Transcripts per million, scaled by length and also suitable for plotting.
- [`raw_counts.csv.gz`](`r file.path(results_dir, "raw_counts.csv.gz")`): Only use to perform a new differential expression analysis. These counts will vary across samples due to differences in sequencing depth, and have not been normalized. Do not use this file for plotting genes.


# MultiQC

Results from FastQC and other metrics can be found [here](multiqc_report.html)

These reports show per base sequence count irregularities in OCP_CTL1_run1, and to a less degreee, OCP_CTL2_run1, OCP_ZA2_run1 and BM_CTL2_run2, with patterns that indicate possible resequencing of the same sequence. Samples BM_CTL1_run1, BM_CTL3_run1, BM_ZA1_run1, OCP_CTL3_run1, and OCP_ZA1_run1 show a nonrandom distribution of sequence counts, with certain bases (like G and C) showing a higher frequency than expected.

# Read metrics {.tabset}

## Total reads

```{r plot_total_reads}
plotTotalReads(bcb)
```


## Mapped reads

The number of mapped reads should correspond to the number of total reads.

```{r plot_mapped_reads}
plotMappedReads(bcb)
```

Mapped reads are below 5 million in OCP_CTL4_run2, OCP_CTL6_run2, OCP_CTL5_run2, OCP_ZA4_run2

## Mapping rate

The genomic mapping rate represents the percentage of reads mapping to the reference genome. Low mapping rates are indicative of sample contamination, poor sequencing quality or other artifacts.

```{r plot_mapping_rate}
plotMappingRate(bcb)
```

Mapping rates are too low in BM_BOTH2_run1 and OCP_BOTH2_run2 and marginal in OCP_CTL1_run1
## Number of genes detected

```{r plot_genes_detected}
plotGenesDetected(bcb)
```


## Gene detection saturation

We should observe a linear trend in the number of genes detected with the number of mapped reads, which indicates that the sample input was not overloaded.

```{r plot_gene_saturation}
plotGeneSaturation(bcb)
```


## Exonic mapping rate

Ideally, at least 60% of total reads should map to exons.

```{r plot_exonic_mapping_rate}
plotExonicMappingRate(bcb)
```

Low exonic mapping rates in BM_CTL3_run1, BM_ZA1_run1, OCP_CTL3_run1, OCP_ZA1_run1, OCP_ZA2_run1, roughly corresponding to the samples with  biased sequence representation in the MultiQC results.

## Intronic mapping rate

The majority of reads should map to exons and not introns.

```{r plot_intronic_mapping_rate}
plotIntronicMappingRate(bcb)
```


## rRNA mapping rate

Samples should have a ribosomal RNA (rRNA) contamination rate below 10%.

```{r plot_rrna_mapping_rate}
plotRRNAMappingRate(bcb)
```

High rRNA rates in BM_CTL3_run1, BM_ZA1_run1, OCP_CTL3_run1, OCP_ZA1_run1, OCP_ZA2_run1, corresponding to poor exonic rates and biased sequence representation in the MultiQC results. 

## 5'->3' bias

```{r plot_53_bias}
plot53Bias(bcb)
```

## Counts per gene

Generally, we expect similar count spreads for all genes between samples unless the library sizes or total RNA expression are different. The log10 TMM-normalized counts per gene normalization method [@Robinson:2010dd] equates the overall expression levels of genes between samples under the assumption that the majority of them are not differentially expressed. Therefore, by normalizing for total RNA expression by sample, we expect the spread of the log10 TMM-normalized counts per gene to be similar for every sample.

```{r plot_counts_per_gene}
plotCountsPerGene(bcb)
```


## Count density

Generally, we expect similar count spreads for all genes between samples unless the total expressed RNA per sample is different.

```{r plot_count_density}
plotCountDensity(bcb)
```


# Fit modeling

Several quality metrics are first assessed to explore the fit of the model, before differential expression analysis is performed.


## Variance stabilization

The plots below show the standard deviation of normalized counts (`normalized_counts`) using `log2()`, `rlog()`, and variance stabilizing (`vst()`) transformations by `rank(mean)`. The transformations greatly reduce the standard deviation, with `rlog()` stabilizing the variance best across the mean.

```{r plot_mean_sd, fig.height=18, fig.width=6}
plotMeanSD(bcb, orientation = "vertical")
```


## Dispersion

The following plot shows the dispersion by mean of normalized counts. We expect the dispersion to decrease as the mean of normalized counts increases.

```{r plot_disp_ests}
plotDispEsts(bcb)
```



# Sample similarity analysis

Before performing similarity analysis, we transform counts to log2, which acts to minimize large differences in sequencing depth and helps normalize all samples to a similar dynamic range. For RNA-seq count data, variance increases with the mean. Logarithmic transformation of normalized count values with a small pseudocount will account for large variations seen between the highest expressing genes so that these genes won't dominate the PCA plots. However, due to the strong noise among low count values due to Poisson, the general log2 transformation will amplify this noise, and instead, low count genes will now dominate the PCA plots. So instead, we use a transformation (`rlog()`) that gives similar results for high counts as a log2 transformation but also shrinks the values of low counts towards the genesâ€™ average across samples. We do this with the `rlog()` function in the [DESeq2][] package [@DESeq2], which we will later use for differential gene expression analysis.


## Principal component analysis (PCA)

PCA [@Jolliffe:2002wx] is a multivariate technique that allows us to summarize the systematic patterns of variations in the data. PCA takes the expression levels for genes and transforms it in principal component space, reducing each sample into one point. Thereby, we can separate samples by expression variation, and identify potential sample outliers. The PCA plot is a way to look at how samples are clustering.

```{r plot_pca}
plotPCA(bcb, color=NULL)
plotPCA(bcb, label=TRUE, color=NULL)
```

Drop samples with low mapping rates (BM_BOTH2_run1, OCP_BOTH2_run2) which are clear outliers on this plot.

```{r plot_pca_sub1}
mycoldata <- as.data.frame(colData(bcb))

bcb_sub <- bcb[,which(!mycoldata$sampleID %in% c("BM_BOTH2_run1", "OCP_BOTH2_run2"))]

plotPCA(bcb_sub, color=NULL, interestingGroups=c("run", "cellPopulation", "treatment"), label=TRUE)
plotPCA(bcb_sub, color=NULL, label=TRUE)
```

This looks suspiciously like the majority of the resequenced samples were BM samples, and not OCP samples. If that was the case, PC1 would separate the cell types, and PC2 would capture the treatment effects in BM.

I mapped the expression levels of  OCP markers from this [paper](http://www.pnas.org/content/110/30/12385), including Cd27, c-Kit and Cd115 (Csf1r) and by and large (teh markers are protein, so may not perfectly correspond)

```{r switch_check}
tpm_sub <- counts(bcb_sub, normalized = "tpm")
mycoldata <- as.data.frame(colData(bcb_sub))
mycoldata$cd27 <- tpm_sub["ENSMUSG00000030336",]

pcadata <- plotPCA(bcb_sub, interestingGroups=c("run", "treatment", "cellPopulation", "repeated"), returnData=TRUE)
pcadata$cd27 <- tpm_sub["ENSMUSG00000030336",]
pcadata$csf1r <- tpm_sub["ENSMUSG00000024621",]

pcadata$ckit <- tpm_sub["ENSMUSG00000005672",]

library(ggrepel)

ggplot(pcadata, aes(x=pc1,y=pc2,  label=name, color=cd27 ))+geom_point(size=5,  stroke=2)+geom_text_repel(color="darkgrey")


ggplot(pcadata, aes(x=pc1,y=pc2,  label=name, color=csf1r ))+geom_point(size=5,  stroke=2)+geom_text_repel(color="darkgrey")

ggplot(pcadata, aes(x=pc1,y=pc2,  label=name, color=ckit ))+geom_point(size=5,  stroke=2)+geom_text_repel(color="darkgrey")

pcadata <- cbind(pcadata, metadata(bcb_sub)$metrics)


ggplot(pcadata, aes(x=pc1,y=pc2,  label=name, color=totalReads ))+geom_point(size=5,  stroke=2)+geom_text_repel(color="darkgrey")+  theme(legend.position="right")+scale_color_viridis()

ggplot(pcadata, aes(x=pc1,y=pc2,  label=name, color=rrnaRate ))+geom_point(size=5,  stroke=2)+geom_text_repel(color="darkgrey")+  theme(legend.position="right")+scale_color_viridis()

ggplot(pcadata, aes(x=pc1,y=pc2,  label=name, color=x5x3Bias ))+geom_point(size=5,  stroke=2)+geom_text_repel(color="darkgrey")+  theme(legend.position="right")+scale_color_viridis()
```

It looks like the cluster on the right are the original run1 OCP samples. What happens when we zoom in on those.

```{r zoompca}
ggplot(subset(pcadata, pc1>30& pc2>-5), aes(x=pc1,y=pc2,  label=name, color=treatment,shape=run))+geom_point(size=5,  stroke=2)+geom_text_repel(color="darkgrey")
```

```{r run1}
mycoldata <- as.data.frame(colData(bcb_sub))
bcb_sub1 <- bcb_sub[,which(mycoldata$run==1)]
plotPCA(bcb_sub1)
bcb_sub1ocp <- bcb_sub[,which(mycoldata$run==1 & mycoldata$cellPopulation=="OCP")]
 pcadata_sub1ocp <- plotPCA(bcb_sub1ocp, interestingGroups="treatment", returnData=TRUE)

pcadata_sub1ocp <- cbind(pcadata_sub1ocp, metadata(bcb_sub1ocp)$metrics)

ggplot(pcadata_sub1ocp, aes(x=pc1,y=pc2,  label=name, color=exonicRate ))+geom_point(size=5,  stroke=2)+geom_text_repel(color="darkgrey")+  theme(legend.position="right")+scale_color_viridis()
ggplot(pcadata_sub1ocp, aes(x=pc1,y=pc2,  label=name, color=rrnaRate ))+geom_point(size=5,  stroke=2)+geom_text_repel(color="darkgrey")+  theme(legend.position="right")+scale_color_viridis()
ggplot(pcadata_sub1ocp, aes(x=pc1,y=pc2,  label=name, color=mappedReads ))+geom_point(size=5,  stroke=2)+geom_text_repel(color="darkgrey")+  theme(legend.position="right")+scale_color_viridis()

```

```{r dropnonprotein}
counts_sub1ocp <- counts(bcb_sub1ocp, normalized = FALSE)
rowData <- rowData(bcb_sub1ocp)
annotated <- cbind(as.data.frame(counts_sub1ocp), as.data.frame(rowData))


bcb_sub1ocp_protein <- bcb_sub1ocp[which(annotated[,"biotype"]=="protein_coding"|annotated[,"biotype"]=="lincRNA"),]
plotPCA(bcb_sub1ocp_protein, interestingGroups=c("treatment"))
plotPCA(bcb_sub1ocp_protein, interestingGroups=c("rin"))
plotPCA(bcb_sub1ocp_protein, interestingGroups=c("x53Bias_factor"))

bcb_sub <- bcb_sub[,which(colData(bcb_sub)$sampleID!="lung_normal_NA_6890_2")]
```

## Covariates correlation with PCs

When multiple factors may influence the results of a given experiment, it is useful to assess which of them is responsible for the most variance as determined by PCA. We adapted the method described by Daily et al. where they integrated a method to correlate covariates with principal components values to determine the importance of each factor.

Here we are showing the correlational analysis of the rlog transformed count data's principal components with the metadata covariates of interest. Significant correlations (FDR < 0.1) are shaded from blue (anti-correlated) to orange (correlated), with non-significant correlations shaded in gray.

```{r plot_pca_covariates}
plotPCACovariates(bcb_sub)
```


## Hierarchical clustering

Inter-correlation analysis (ICA) is another way to look at how well samples cluster by plotting the correlation between the expression profiles of the samples.

```{r plot_correlation_heatmap}
plotCorrelationHeatmap(bcb)
```



```{r footer, child="_footer.Rmd"}
```
